<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Stephan Schiffels">
<meta name="dcterms.date" content="2023-10-26">

<title>Computational Methods for human population genetics and ancient DNA - 3&nbsp; Introduction to F3- and F4-Statistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./eager.html" rel="next">
<link href="./poseidon.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./fstats.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Computational Methods for human population genetics and ancient DNA</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Quarto_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to Quarto</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./poseidon.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fstats.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eager.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to nf-core/eager</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mobest.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">mobest - Spatiotemporal Ancestry Interpolation and Search</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fst.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Measuring population structure using Fst</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pca_mds.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Dimensionality reduction using PCA and MDS</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pmrread.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Pairwise mismatch rate (PMR) and READ relatedness inference</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#admixture---f3-statistics" id="toc-admixture---f3-statistics" class="nav-link active" data-scroll-target="#admixture---f3-statistics"><span class="header-section-number">3.1</span> Admixture - F3 Statistics</a>
  <ul class="collapse">
  <li><a href="#computing-admixture-f3-with-xerxes" id="toc-computing-admixture-f3-with-xerxes" class="nav-link" data-scroll-target="#computing-admixture-f3-with-xerxes"><span class="header-section-number">3.1.1</span> Computing Admixture-F3 with xerxes</a></li>
  <li><a href="#running-xerxes-via-a-configuration-file" id="toc-running-xerxes-via-a-configuration-file" class="nav-link" data-scroll-target="#running-xerxes-via-a-configuration-file"><span class="header-section-number">3.1.2</span> Running xerxes via a configuration file</a></li>
  </ul></li>
  <li><a href="#f4-statistics" id="toc-f4-statistics" class="nav-link" data-scroll-target="#f4-statistics"><span class="header-section-number">3.2</span> F4 Statistics</a>
  <ul class="collapse">
  <li><a href="#shaping-intuition---the-abba--and-baba-sites" id="toc-shaping-intuition---the-abba--and-baba-sites" class="nav-link" data-scroll-target="#shaping-intuition---the-abba--and-baba-sites"><span class="header-section-number">3.2.1</span> Shaping intuition - the ABBA- and BABA-sites</a></li>
  <li><a href="#from-single-samples-to-allele-frequencies" id="toc-from-single-samples-to-allele-frequencies" class="nav-link" data-scroll-target="#from-single-samples-to-allele-frequencies"><span class="header-section-number">3.2.2</span> From single samples to allele frequencies</a></li>
  <li><a href="#running-f4-statistics-with-xerxes" id="toc-running-f4-statistics-with-xerxes" class="nav-link" data-scroll-target="#running-f4-statistics-with-xerxes"><span class="header-section-number">3.2.3</span> Running F4-Statistics with xerxes</a></li>
  </ul></li>
  <li><a href="#outgroup-f3-statistics" id="toc-outgroup-f3-statistics" class="nav-link" data-scroll-target="#outgroup-f3-statistics"><span class="header-section-number">3.3</span> Outgroup-F3-Statistics</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Stephan Schiffels </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 26, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="admixture---f3-statistics" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="admixture---f3-statistics"><span class="header-section-number">3.1</span> Admixture - F3 Statistics</h2>
<p>F3 statistics are a useful analytical tool to understand population relationships. F3 statistics, just as F4 and F2 statistics measure allele frequency correlations between populations and were introduced by Nick Patterson <span class="citation" data-cites="Patterson2012">(<a href="references.html#ref-Patterson2012" role="doc-biblioref">Patterson et al. 2012</a>)</span>, but see also <span class="citation" data-cites="Peter2016">(<a href="references.html#ref-Peter2016" role="doc-biblioref">Peter 2016</a>)</span> for another introduction.</p>
<p>F3 statistics are used for two purposes: i) as a test whether a target population (C) is admixed between two source populations (A and B), and ii) to measure shared drift between two test populations (A and B) from an outgroup (C).</p>
<p>F3 statistics are in both cases defined as the product of allele frequency differences between population C to A and B, respectively:</p>
<p><span class="math display">\[F3(A,B;C)=\langle(c−a)(c−b)\rangle\]</span></p>
<p>Here, <span class="math inline">\(\langle\cdot\rangle\)</span> denotes the average over all genotyped sites, and a, b and c denote the allele frequency for a given site in the three populations A, B and C.</p>
<p>It can be shown that if <span class="math inline">\(F3(A, B; C)\)</span> is negative, it provides unambiguous proof that population C is admixed between populations A and B, as in the following phylogeny (taken from Figure 1 from <span class="citation" data-cites="Patterson2012">(<a href="references.html#ref-Patterson2012" role="doc-biblioref">Patterson et al. 2012</a>)</span>:</p>
<p><img src="img/fstats/f3_phylogeny.png" alt="F3-phylogeny" style="width: 300px;"></p>
<p>Intuitively, an F3 statistics becomes negative if the allele frequency of the target population C is on average intermediate between the allele frequencies of A and B. Consider as an extreme example a genomic site where <span class="math inline">\(a=0\)</span>, <span class="math inline">\(b=1\)</span> and <span class="math inline">\(c=0.5\)</span>. Then we have <span class="math inline">\((c−a)(c−b)=−0.25\)</span>, which is negative. So if the entire statistics is negative, it suggests that in many positions, the allele frequency c is indeed intermediate, suggesting admixture between the two sources.</p>
<p>One way to understand this is by looking what happens to a list of SNPs and allele frequencies for groups A, B and C:</p>
<table class="table">
<thead>
<tr class="header">
<th>SNP</th>
<th>A</th>
<th>B</th>
<th>C</th>
<th><span class="math inline">\((c-a)(c-b)\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>1</td>
<td>0</td>
<td>0.5</td>
<td>-0.25</td>
</tr>
<tr class="even">
<td>2</td>
<td>0.8</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>3</td>
<td>0</td>
<td>0.7</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>4</td>
<td>0.1</td>
<td>0.5</td>
<td>0.3</td>
<td>-0.04</td>
</tr>
<tr class="odd">
<td>5</td>
<td>0</td>
<td>0.1</td>
<td>0.2</td>
<td>0.02</td>
</tr>
<tr class="even">
<td>6</td>
<td>1</td>
<td>0.2</td>
<td>0.9</td>
<td>-0.07</td>
</tr>
<tr class="odd">
<td><span class="math inline">\(F_3(A, B;C)\)</span></td>
<td></td>
<td></td>
<td></td>
<td><strong>-0.057</strong></td>
</tr>
</tbody>
</table>
<p>Every SNP where C has an allele frequency <em>intermediate</em> between A and B contributes negatively. Here, the average is also negative, providing evidence for admixture. For statistical certainty, an error bar for this estimate is needed, which is typically computed via Jackknife (see for example the <a href="https://github.com/poseidon-framework/poseidon-analysis-hs/blob/main/docs/xerxes_whitepaper.pdf">xerxes whitepaper</a>).</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>If an F3 statistics is <em>not</em> negative, it does <em>not</em> proof that there is no admixture!</p>
</div>
</div>
<section id="computing-admixture-f3-with-xerxes" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="computing-admixture-f3-with-xerxes"><span class="header-section-number">3.1.1</span> Computing Admixture-F3 with xerxes</h3>
<p>We will use this statistics to test if Finnish are admixed between East and West, using different Eastern and Western sources. In the West, we use French, Icelandic, Lithuanian and Norwegian as source, and in the East we use Nganasan and one of the ancient individuals analysed in <span class="citation" data-cites="Lamnidis2018">(<a href="references.html#ref-Lamnidis2018" role="doc-biblioref">Lamnidis et al. 2018</a>)</span>, from the site of <em>Bolshoy Oleni Ostrov</em> from the Northern Russian Kola-peninsula, and dating to 3500 years before present.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you happen to have downloaded a copy of the <a href="https://www.poseidon-adna.org/#/archive_overview?id=public-poseidon-archives">Poseidon Community Archive</a> already, then just use the path to that archive in the following commands. Otherwise you download the entire archive via</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trident</span> fetch <span class="at">-d</span> /somewhere/to/store/the/archive <span class="at">--downloadAll</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>or just the relevant packages for the examples in this chapter:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trident</span> fetch <span class="at">-d</span> /somewhere/to/store/the/archive <span class="at">-f</span> <span class="st">"*2012_PattersonGenetics*,*2014_RaghavanNature*,*2014_LazaridisNature*,*2018_Lamnidis_Fennoscandia*"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>We use the software <code>xerxes fstats</code> from the <a href="https://www.poseidon-adna.org/#/xerxes?id=xerxes-cli-software">Poseidon Framework</a>. Here is a command line that computes 8 statistics for us:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">xerxes</span> fstats <span class="at">-d</span> ~/dev/poseidon-framework/community-archive <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Nganasan,French,Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Nganasan, Icelandic, Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Nganasan, Lithuanian, Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Nganasan, Norwegian, Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Russia_Bolshoy, French, Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Russia_Bolshoy, Icelandic, Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Russia_Bolshoy, Lithuanian, Finnish)"</span> <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--stat</span> <span class="st">"F3(Russia_Bolshoy, Norwegian, Finnish)"</span> <span class="dt">\</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that <code>xerxes fstats</code> will automatically find the right packages from your local archive that contain these groups. You can see in the output of the program which packages contribute:</p>
<pre><code>[Info]    5 relevant packages for chosen statistics identified:
[Info]    *2012_PattersonGenetics-2.1.3*
[Info]    *2014_LazaridisNature-4.0.2*
[Info]    *2016_LazaridisNature-2.1.3*
[Info]    *2018_Lamnidis_Fennoscandia-2.1.0*
[Info]    *2019_Flegontov_PalaeoEskimo-2.2.1*</code></pre>
<p>So these five packages contain the samples requested in these statistics. You can inquire about this also more manually using <a href="https://www.poseidon-adna.org/#/trident?id=list-command"><code>trident list</code></a></p>
</div>
</div>
<p>Here is the result that you should get, nicely layouted in a Text-table:</p>
<pre><code>.-----------.----------------.------------.---------.---.---------.----------------.--------------------.------------------.---------------------.
| Statistic |       a        |     b      |    c    | d | NrSites | Estimate_Total | Estimate_Jackknife | StdErr_Jackknife |  Z_score_Jackknife  |
:===========:================:============:=========:===:=========:================:====================:==================:=====================:
| F3        | Nganasan       | French     | Finnish |   | 593124  | -1.0450e-3     | -1.0451e-3         | 1.2669e-4        | -8.249133659451905  |
| F3        | Nganasan       | Icelandic  | Finnish |   | 593124  | -1.1920e-3     | -1.1920e-3         | 1.3381e-4        | -8.908381869946188  |
| F3        | Nganasan       | Lithuanian | Finnish |   | 593124  | -1.1605e-3     | -1.1605e-3         | 1.5540e-4        | -7.4680182465607245 |
| F3        | Nganasan       | Norwegian  | Finnish |   | 593124  | -1.0913e-3     | -1.0914e-3         | 1.3921e-4        | -7.83945981272796   |
| F3        | Russia_Bolshoy | French     | Finnish |   | 542789  | -6.1807e-4     | -6.1809e-4         | 1.0200e-4        | -6.059872900228102  |
| F3        | Russia_Bolshoy | Icelandic  | Finnish |   | 542789  | -6.2801e-4     | -6.2802e-4         | 1.1792e-4        | -5.325695961373772  |
| F3        | Russia_Bolshoy | Lithuanian | Finnish |   | 542789  | -3.7310e-4     | -3.7310e-4         | 1.2973e-4        | -2.8760029685791637 |
| F3        | Russia_Bolshoy | Norwegian  | Finnish |   | 542789  | -3.7646e-4     | -3.7653e-4         | 1.1630e-4        | -3.2375830440434323 |
'-----------'----------------'------------'---------'---'---------'----------------'--------------------'------------------'---------------------'</code></pre>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the option <code>-f &lt;FILE&gt;</code> to output the results additionally to a tab-separated file, or <code>--raw</code> if you prefer the standard output to be tab-separated</p>
</div>
</div>
<p>You can see that in all cases this statistic is negative (<code>Estimate_Total</code>). The next two columns ( <code>Estimate_Jackknife</code> and <code>StdErr_Jackknife</code>) are computed using Jackknifing (see <a href="https://github.com/poseidon-framework/poseidon-analysis-hs/blob/main/docs/xerxes_whitepaper.pdf">the xerxes whitepaper</a> for details). The last column is the Z score, and it is important here: It gives the deviation of the f3 statistic from zero in units of the standard error. As general rule, a Z score of -3 or more suggests a significant rejection of the Null hypothesis that the statistic is not negative. In this case, all of the statistics are significantly negative (with one borderline exception), proving that Finnish have ancestral admixture of East and West Eurasian ancestry. Here, Eastern ancestry is represented by Nganasan or Russia_Bolshoy, respectively, while Western ancestry by French, Icelandic, Lithuanian and Norwegian, respectively. Note that the statistics does not suggest when this admixture happened!</p>
</section>
<section id="running-xerxes-via-a-configuration-file" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="running-xerxes-via-a-configuration-file"><span class="header-section-number">3.1.2</span> Running xerxes via a configuration file</h3>
<p>As you will have noticed, the command line above is getting quite long, since a separate <code>--stat</code> option has to be entered for every statistic to be computed. There is a more powerful and elegant interface to xerxes, which uses a configuration file in <a href="https://de.wikipedia.org/wiki/YAML">YAML format</a>. To illustrate it, let us consider the configuration file (which can be found in <code>fstats_working/F3_finnish.config</code> in the git-repository of this book) needed to compute the same statistic as above:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fstats</span><span class="kw">:</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="kw">-</span><span class="at"> </span><span class="fu">type</span><span class="kw">:</span><span class="at"> F3</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">a</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Nganasan"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Russia_Bolshoy"</span><span class="kw">]</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">b</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"French"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Icelandic"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Lithuanian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Norwegian"</span><span class="kw">]</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">c</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Finnish"</span><span class="kw">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can then run xerxes as</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">xerxes</span> fstats <span class="at">-d</span> ~/dev/poseidon-framework/community-archive <span class="at">--statConfig</span> fstats_working/F3_finnish.config</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So <code>xerxes</code> then automatically creates all combinations of populations listed in slots <code>a</code>, <code>b</code> and <code>c</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that there are actually three types of F3-statistics supported by xerxes:</p>
<ul>
<li><code>F3vanilla</code>: The purest form, defined literally as <span class="math inline">\(\langle(c−a)(c−b)\rangle\)</span></li>
<li><code>F3</code>: A bias-corrected version, which is only valid for groups in C that have non-zero heterozygosity</li>
<li><code>F3star</code>: This one is normalised by the heterozgygosity of the third population, C, as suggested in <span class="citation" data-cites="Patterson2012">(<a href="references.html#ref-Patterson2012" role="doc-biblioref">Patterson et al. 2012</a>)</span> and implemented in the <a href="https://github.com/DReichLab/AdmixTools">Admixtools package</a>.</li>
</ul>
<p>The <a href="(https://github.com/poseidon-framework/poseidon-analysis-hs/blob/main/docs/xerxes_whitepaper.pdf)">white-paper</a> explains this in detail.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The configuration file format has a lot more options. Here is a bit more complex example, but see also <a href="https://www.poseidon-adna.org/#/xerxes?id=input-via-a-configuraton-file">the documentation</a>:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You can define groups right within the configuration file.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># here we use negative selection to remove individuals from the</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># newly defined groups</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">groupDefs</span><span class="kw">:</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">CEU2</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"CEU.SG"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;NA12889.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;NA12890.SG&gt;"</span><span class="kw">]</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">FIN2</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"FIN.SG"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;HG00383.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;HG00384.SG&gt;"</span><span class="kw">]</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">GBR2</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"GBR.SG"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;HG01791.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;HG02215.SG&gt;"</span><span class="kw">]</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">IBS2</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"IBS.SG"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;HG02238.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"-&lt;HG02239.SG&gt;"</span><span class="kw">]</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="fu">fstats</span><span class="kw">:</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="kw">-</span><span class="at"> </span><span class="fu">type</span><span class="kw">:</span><span class="at"> F2</span><span class="co"> # this will create 2x2 = 4 F2-Statistics</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">a</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"French"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Spanish"</span><span class="kw">]</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">b</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Han"</span><span class="kw">,</span><span class="at"> </span><span class="st">"CEU2"</span><span class="kw">]</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="kw">-</span><span class="at"> </span><span class="fu">type</span><span class="kw">:</span><span class="at"> F3vanilla</span><span class="co"> # This will create 3x2x1 = 6 Statistics</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">a</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"French"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Spanish"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Mbuti"</span><span class="kw">]</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">b</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Han"</span><span class="kw">,</span><span class="at"> </span><span class="st">"CEU2"</span><span class="kw">]</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">c</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"&lt;Chimp.REF&gt;"</span><span class="kw">]</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="kw">-</span><span class="at"> </span><span class="fu">type</span><span class="kw">:</span><span class="at"> F4</span><span class="co"> # This will create 5x5x4x1 = 100 Statistics</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">a</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"&lt;I0156.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0157.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0159.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0160.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0161.SG&gt;"</span><span class="kw">]</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">b</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"&lt;I0156.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0157.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0159.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0160.SG&gt;"</span><span class="kw">,</span><span class="at"> </span><span class="st">"&lt;I0161.SG&gt;"</span><span class="kw">]</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">c</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"CEU2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"FIN2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"GBR2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"IBS2"</span><span class="kw">]</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">d</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"&lt;Chimp.REF&gt;"</span><span class="kw">]</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Altogether: 110 statistics of different types</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>which will not just create multiple statistic using row-combinations, as described, but also uses newly defined groups and combines multiple statistic types (F2, F3 and F4) in one run.</p>
</div>
</div>
</section>
</section>
<section id="f4-statistics" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="f4-statistics"><span class="header-section-number">3.2</span> F4 Statistics</h2>
<p>A different way to test for admixture is by “F4 statistics” (or “D statistics” which is very similar), also introduced in <span class="citation" data-cites="Patterson2012">(<a href="references.html#ref-Patterson2012" role="doc-biblioref">Patterson et al. 2012</a>)</span>.</p>
<p>F4 statistics are also defined in terms of correlations of allele frequency differences, similarly to F3 statistics, but involving four different populations, not just three. Specifically we define</p>
<p><span class="math display">\[F4(A,B;C,D)=\langle(a−b)(c−d)\rangle.\]</span></p>
<section id="shaping-intuition---the-abba--and-baba-sites" class="level3" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="shaping-intuition---the-abba--and-baba-sites"><span class="header-section-number">3.2.1</span> Shaping intuition - the ABBA- and BABA-sites</h3>
<p>To understand the statistics, consider the following tree:</p>
<p><img src="img/fstats/f4_phylogeny.png" alt="F4-phylogeny" style="width: 300px;"></p>
<p>In this tree, without any additional admixture, the allele frequency difference between A and B should be completely independent from the allele frequency difference between C and D. In that case, F4(A, B; C, D) should be zero, or at least not statistically different from zero. However, if there was gene flow from C or D into A or B, the statistic should be different from zero. Specifically, if the statistic is significantly negative, it implies gene flow between either C and B, or D and A. If it is significantly positive, it implies gene flow between A and C, or B and D.</p>
<p>It is helpful to again consider an example using a SNP list, this time assuming that every population is just a single (haploid) individual, so each allele frequency can just be 0 or 1. For example:</p>
<table class="table">
<thead>
<tr class="header">
<th>SNP</th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>D</th>
<th><span class="math inline">\((a-b)(c-d)\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>2</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0</td>
</tr>
<tr class="odd">
<td>3</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>-1</td>
</tr>
<tr class="even">
<td>4</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="odd">
<td>5</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>-1</td>
</tr>
<tr class="even">
<td>6</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td><span class="math inline">\(F_4(A, B;C, D)\)</span></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td><strong>-0.0167</strong></td>
</tr>
</tbody>
</table>
<p>You can see that the only SNPs that contribute positively to this statistics are SNPs where the alleles are distributed as <code>1010</code> and <code>0101</code>, and the only SNPs that contribute negatively are <code>1001</code> and <code>0110</code>. In the literature, the two patterns have been dubbed “ABBA” and “BABA”, which is why the statistical test behind this statistic (see below) was sometimes called the ABBA-BABA test (see for example <span class="citation" data-cites="Martin2015">(<a href="references.html#ref-Martin2015" role="doc-biblioref">Martin, Davey, and Jiggins 2015</a>)</span>).</p>
<p>The intuition here is straight-forward: In positions that are polymorphic in both <span class="math inline">\((A,B)\)</span> and <span class="math inline">\((C,D)\)</span>, this statistic asks whether B is genetically more similar to C than it is to D. This is most useful as a test for “treeness”: If A, B, C, D are related to each other as indicated in the above tree, then C should be equally closely related to C as to D. But if we actually find evidence that B is closer to C than to D, or vice versa, then this means that the tree above cannot be correct, but that there must be a closer connection between B and C or B and D, depending on the sign of the statistic.</p>
</section>
<section id="from-single-samples-to-allele-frequencies" class="level3" data-number="3.2.2">
<h3 data-number="3.2.2" class="anchored" data-anchor-id="from-single-samples-to-allele-frequencies"><span class="header-section-number">3.2.2</span> From single samples to allele frequencies</h3>
<p>So the ABBA- and BABA-categories of SNPs help shape intuition for how this statistic behaves for single haploid genomes. But what about population allele frequencies? Looking back at the formula <span class="math inline">\(\langle(a−b)(c−d)\rangle\)</span> this doesn’t help very much with intuition how this behaves with frequencies. Well, a nice feature of F4-Statistics is that averages factor out. This means, that if you have multiple samples in one or multiple slots A, B, C or D, the total F4-statistic of the <em>groups</em> is exactly equal to the <em>average</em> of F4-Statistics of the <em>individuals</em>. Here is a more mathematical definition.</p>
<p>Let’s say we have 2 individuals in each of A and B, so we may perhaps write <span class="math inline">\(A=\{A_1,A_2\}\)</span> and <span class="math inline">\(B=\{B_1,B_2\}\)</span>. Then one can show to have</p>
<p><span class="math display">\[F4(A, B; C, D) = \text{Average of}[F4(A_1, B_1; C, D), F4(A_1, B_2; C, D), F4(A_2, B_1; C, D), F4(A_2, B_2; C, D)]\]</span></p>
<p>so just thte average over all individual-based F4-statistics. And this can be shown to be true for arbitrary numbers of samples. So in other words: An F4-Statistic <em>always</em> measures the <em>average</em> excess of pairwise BABA SNPs over ABBA SNPs. To me, this is a useful insight, as I find thinking in terms of ABBA-BABA somehow more helpful than thinking in terms of correlations of allele-frequency differences (which is really what the original formula is).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>F4-statistics have been famously used to show that Neanderthals are more closely related to Non-African populations than to Africans, suggesting gene-flow between Neanderthals and Non-Africans (shown in <span class="citation" data-cites="Green2010">(<a href="references.html#ref-Green2010" role="doc-biblioref">Green et al. 2010</a>)</span>). You can reproduce this famous result with</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">xerxes</span> fstats <span class="at">-d</span> ~/poseidon_repo <span class="at">--stat</span> <span class="st">'F4(&lt;Chimp.REF&gt;,&lt;Altai_published.DG&gt;,Yoruba,French)'</span> <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--stat</span> <span class="st">'F4(&lt;Chimp.REF&gt;,&lt;Altai_published.DG&gt;,Sardinian,French)'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>which shows that the first statistic is significantly positive with a Z-score of 7.99, while the second one is insignificantly different from zero (Z=1.01)</p>
</div>
</div>
<p>The way this statistic is often used, is to put a divergent outgroup as population A, for which we know for sure that there was no admixture into either C or D. With this setup, we can then test for gene flow between B and D (if the statistic is positive), or B and C (if it is negative).</p>
</section>
<section id="running-f4-statistics-with-xerxes" class="level3" data-number="3.2.3">
<h3 data-number="3.2.3" class="anchored" data-anchor-id="running-f4-statistics-with-xerxes"><span class="header-section-number">3.2.3</span> Running F4-Statistics with xerxes</h3>
<p>Here, we can use this statistic to test for East Asian admixture in Finns, similarly to the test using Admixture F3 statistics above. We will again use <code>xerxes fstats</code>. We again prepare a configuration file (in <code>fstats_working/F4_finish.config</code> in the git-repository of this book), this time with four populations (A, B, C, D):</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fstats</span><span class="kw">:</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="kw">-</span><span class="at"> </span><span class="fu">type</span><span class="kw">:</span><span class="at"> F4</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">a</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Mbuti"</span><span class="kw">]</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">b</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Nganasan"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Russia_Bolshoy"</span><span class="kw">]</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">c</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"French"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Icelandic"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Lithuanian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Norwegian"</span><span class="kw">]</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">d</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Finnish"</span><span class="kw">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can again run via</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">xerxes</span> fstats <span class="at">-d</span> ~/dev/poseidon-framework/community-archive <span class="at">--statConfig</span> fstats_working/F4_finnish.config</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The result is:</p>
<pre><code>.-----------.-------.----------------.------------.---------.---------.----------------.--------------------.------------------.--------------------.
| Statistic |   a   |       b        |     c      |    d    | NrSites | Estimate_Total | Estimate_Jackknife | StdErr_Jackknife | Z_score_Jackknife  |
:===========:=======:================:============:=========:=========:================:====================:==================:====================:
| F4        | Mbuti | Nganasan       | French     | Finnish | 593124  | 2.3114e-3      | 2.3115e-3          | 1.2676e-4        | 18.235604067907143 |
| F4        | Mbuti | Nganasan       | Icelandic  | Finnish | 593124  | 1.6590e-3      | 1.6590e-3          | 1.4861e-4        | 11.163339072181776 |
| F4        | Mbuti | Nganasan       | Lithuanian | Finnish | 593124  | 1.3290e-3      | 1.3290e-3          | 1.4681e-4        | 9.052979707622278  |
| F4        | Mbuti | Nganasan       | Norwegian  | Finnish | 593124  | 1.6503e-3      | 1.6503e-3          | 1.5358e-4        | 10.745850997260929 |
| F4        | Mbuti | Russia_Bolshoy | French     | Finnish | 542789  | 1.8785e-3      | 1.8785e-3          | 1.2646e-4        | 14.854487416366263 |
| F4        | Mbuti | Russia_Bolshoy | Icelandic  | Finnish | 542789  | 1.0829e-3      | 1.0828e-3          | 1.4963e-4        | 7.236818881873822  |
| F4        | Mbuti | Russia_Bolshoy | Lithuanian | Finnish | 542789  | 5.4902e-4      | 5.4907e-4          | 1.4601e-4        | 3.7605973064589096 |
| F4        | Mbuti | Russia_Bolshoy | Norwegian  | Finnish | 542789  | 9.3473e-4      | 9.3475e-4          | 1.5302e-4        | 6.108881868125652  |
'-----------'-------'----------------'------------'---------'---------'----------------'--------------------'------------------'--------------------'</code></pre>
<p>As you can see, in all cases, the Z score is positive and larger than 3, indicating a significant deviation from zero, and implying gene flow between Nganasan and Finnish, and BolshoyOleniOstrov and Finnish, when compared to French, Icelandic, Lithuanian or Norwegian.</p>
</section>
</section>
<section id="outgroup-f3-statistics" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="outgroup-f3-statistics"><span class="header-section-number">3.3</span> Outgroup-F3-Statistics</h2>
<p>Outgroup F3 statistics are a special case how to use F3 statistics. The definition is the same as for Admixture F3 statistics, but instead of a target C and two source populations A and B, one now gives an outgroup C and two test populations A and B.</p>
<p>To get an intuition for this statistics, consider the following tree:</p>
<p><img src="img/fstats/outgroupf3_phylogeny.png" alt="Outgroup-F3-phylogeny" style="width: 300px;"></p>
<p>In this scenario, the statistic F3(A, B; C) measures the branch length from C to the common ancestor of A and B, coloured red. So this statistic is simply a measure of how closely two population A and B are related with each other, as measured from a distant outgroup. It is thus a similarity measure: The higher the statistic, the more genetically similar A and B are to one another.</p>
<p>Here is again a SNP table to illustrate, using haploid individuals:</p>
<table class="table">
<thead>
<tr class="header">
<th>SNP</th>
<th>A</th>
<th>B</th>
<th>C</th>
<th><span class="math inline">\((c-a)(c-b)\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>2</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>3</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="even">
<td>4</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>0</td>
</tr>
<tr class="odd">
<td>5</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr class="even">
<td>6</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="odd">
<td><span class="math inline">\(F_3(A, B;C)\)</span></td>
<td></td>
<td></td>
<td></td>
<td><strong>0.5</strong></td>
</tr>
</tbody>
</table>
<p>You can see that each position which is similar between A and B, but different to C contributes 1, all other SNPs 0. So it directly measures similarity between A and B on alleles that differ from the outgroup C.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that the averaging-relation shown for F4 statistics above is also true for Outgroup-F3 statistics, but only for populations A and B, not for C. So if you have multiple samples in A and B, you may think of this statistic being the <em>average</em> over all pairwise nucleotide similarities between individuals in A and B with respect to the same outgroup C.</p>
</div>
</div>
<p>We can use this statistic to measure for example the genetic affinity to East Asia, by performing the statistic F3(Han, X; Mbuti), where Mbuti is a distant African population and acts as outgroup here, Han denote Han Chinese, and X denotes various European populations that we want to test.</p>
<p>You can again define a configuration file that performs looping over various populations X for you:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fstats</span><span class="kw">:</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="kw">-</span><span class="at"> </span><span class="fu">type</span><span class="kw">:</span><span class="at"> F3</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">a</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Han"</span><span class="kw">]</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">b</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Chuvash"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Albanian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Armenian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Bulgarian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Czech"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Druze"</span><span class="kw">,</span><span class="at"> </span><span class="st">"English"</span><span class="kw">,</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="st">"Estonian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Finnish"</span><span class="kw">,</span><span class="at"> </span><span class="st">"French"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Georgian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Greek"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Hungarian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Icelandic"</span><span class="kw">,</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="st">"Italian_North"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Italian_South"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Lithuanian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Maltese"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Mordovian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Norwegian"</span><span class="kw">,</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="st">"Orcadian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Russian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Sardinian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Scottish"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Sicilian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Spanish_North"</span><span class="kw">,</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="st">"Spanish"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Ukrainian"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Finland_Levanluhta"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Russia_Bolshoy"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Russia_Chalmny_Varre"</span><span class="kw">,</span><span class="at"> </span><span class="st">"Saami.DG"</span><span class="kw">]</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">c</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"Mbuti"</span><span class="kw">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>which cycles through many populations from Europe, including the ancient individuals from Chalmny Varre, Bolshoy Oleni Ostrov and Levänluhta (described in <span class="citation" data-cites="Lamnidis2018">(<a href="references.html#ref-Lamnidis2018" role="doc-biblioref">Lamnidis et al. 2018</a>)</span>). We store this file in a file called <code>fstats_working/OutgroupF3_europe.config</code> and run via:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">xerxes</span> fstats <span class="at">--statConfig</span> fstats_working/OutgroupF3_europe.config <span class="at">-d</span> ~/dev/poseidon-framework/community-archive <span class="at">-f</span> fstats_working/outgroupf3_europe.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Often in Outgroup-F3-statistics you use single genomes for population C, sometimes even single haploid genomes. In this case, <code>F3</code> and <code>F3star</code> will get undefined results, because ordinary <code>F3</code> and <code>F3star</code> statistics require population C to have non-zero average heterozygosity, so you will need at least one diploid sample, or multiple haploid or diploid samples.</p>
<p>Use <code>F3vanilla</code> if your third population C is a single pseudo-haploid sample.</p>
</div>
</div>
<p>Here is the output of this run (but note that a tab-separated version was also stored in <code>fstats_working/outgroupf3_europe.tsv</code> using the option <code>-f</code>):</p>
<pre><code>.-----------.-----.----------------------.-------.---.---------.----------------.--------------------.------------------.--------------------.
| Statistic |  a  |          b           |   c   | d | NrSites | Estimate_Total | Estimate_Jackknife | StdErr_Jackknife | Z_score_Jackknife  |
:===========:=====:======================:=======:===:=========:================:====================:==================:====================:
| F3        | Han | Chuvash              | Mbuti |   | 593124  | 5.3967e-2      | 5.3967e-2          | 5.0668e-4        | 106.51180329550319 |
| F3        | Han | Albanian             | Mbuti |   | 593124  | 4.9972e-2      | 4.9973e-2          | 4.9520e-4        | 100.91326321202445 |
| F3        | Han | Armenian             | Mbuti |   | 593124  | 4.9531e-2      | 4.9531e-2          | 4.7771e-4        | 103.68366652942314 |
| F3        | Han | Bulgarian            | Mbuti |   | 593124  | 5.0103e-2      | 5.0103e-2          | 4.8624e-4        | 103.04188532686614 |
| F3        | Han | Czech                | Mbuti |   | 593124  | 5.0536e-2      | 5.0536e-2          | 4.9261e-4        | 102.58792370749681 |
| F3        | Han | Druze                | Mbuti |   | 593124  | 4.8564e-2      | 4.8564e-2          | 4.6788e-4        | 103.79674299622445 |
| F3        | Han | English              | Mbuti |   | 593124  | 5.0280e-2      | 5.0281e-2          | 4.9183e-4        | 102.23198323949656 |
| F3        | Han | Estonian             | Mbuti |   | 593124  | 5.1154e-2      | 5.1155e-2          | 5.0350e-4        | 101.59882496016485 |
| F3        | Han | Finnish              | Mbuti |   | 593124  | 5.1784e-2      | 5.1784e-2          | 5.0603e-4        | 102.33488758899031 |
| F3        | Han | French               | Mbuti |   | 593124  | 5.0207e-2      | 5.0208e-2          | 4.8552e-4        | 103.40976592749682 |
| F3        | Han | Georgian             | Mbuti |   | 593124  | 4.9711e-2      | 4.9711e-2          | 4.8100e-4        | 103.34881140790415 |
| F3        | Han | Greek                | Mbuti |   | 593124  | 4.9874e-2      | 4.9874e-2          | 4.8994e-4        | 101.79554640756365 |
| F3        | Han | Hungarian            | Mbuti |   | 593124  | 5.0497e-2      | 5.0498e-2          | 4.9878e-4        | 101.24215699276706 |
| F3        | Han | Icelandic            | Mbuti |   | 593124  | 5.0680e-2      | 5.0680e-2          | 4.9729e-4        | 101.91303336514295 |
| F3        | Han | Italian_North        | Mbuti |   | 593124  | 4.9903e-2      | 4.9904e-2          | 4.8436e-4        | 103.03094306099203 |
| F3        | Han | Italian_South        | Mbuti |   | 592980  | 4.9201e-2      | 4.9201e-2          | 5.1170e-4        | 96.15239597674244  |
| F3        | Han | Lithuanian           | Mbuti |   | 593124  | 5.0896e-2      | 5.0896e-2          | 5.0638e-4        | 100.50984037418753 |
| F3        | Han | Maltese              | Mbuti |   | 593124  | 4.8751e-2      | 4.8751e-2          | 4.7500e-4        | 102.63442479673623 |
| F3        | Han | Mordovian            | Mbuti |   | 593124  | 5.1820e-2      | 5.1820e-2          | 4.8853e-4        | 106.07409963190884 |
| F3        | Han | Norwegian            | Mbuti |   | 593124  | 5.0724e-2      | 5.0724e-2          | 4.9514e-4        | 102.4454387098217  |
| F3        | Han | Orcadian             | Mbuti |   | 593124  | 5.0469e-2      | 5.0469e-2          | 4.9485e-4        | 101.98814656611475 |
| F3        | Han | Russian              | Mbuti |   | 593124  | 5.1277e-2      | 5.1277e-2          | 4.8613e-4        | 105.48070801791317 |
| F3        | Han | Sardinian            | Mbuti |   | 593124  | 4.9416e-2      | 4.9417e-2          | 4.8908e-4        | 101.04049389691913 |
| F3        | Han | Scottish             | Mbuti |   | 593124  | 5.0635e-2      | 5.0635e-2          | 5.0565e-4        | 100.13962104744425 |
| F3        | Han | Sicilian             | Mbuti |   | 593124  | 4.9194e-2      | 4.9194e-2          | 4.8157e-4        | 102.15353663091187 |
| F3        | Han | Spanish_North        | Mbuti |   | 593124  | 5.0032e-2      | 5.0032e-2          | 4.9377e-4        | 101.32594226555439 |
| F3        | Han | Spanish              | Mbuti |   | 593124  | 4.9693e-2      | 4.9693e-2          | 4.8551e-4        | 102.35200847948641 |
| F3        | Han | Ukrainian            | Mbuti |   | 593124  | 5.0731e-2      | 5.0731e-2          | 4.9506e-4        | 102.47529111692852 |
| F3        | Han | Finland_Levanluhta   | Mbuti |   | 303033  | 5.4488e-2      | 5.4488e-2          | 5.7681e-4        | 94.46487653920919  |
| F3        | Han | Russia_Bolshoy       | Mbuti |   | 542789  | 5.7273e-2      | 5.7273e-2          | 5.2875e-4        | 108.31739594898687 |
| F3        | Han | Russia_Chalmny_Varre | Mbuti |   | 428215  | 5.4000e-2      | 5.4000e-2          | 5.6936e-4        | 94.84371082564112  |
| F3        | Han | Saami.DG             | Mbuti |   | 585193  | 5.4727e-2      | 5.4728e-2          | 5.5546e-4        | 98.5265149143263   |
'-----------'-----'----------------------'-------'---'---------'----------------'--------------------'------------------'--------------------'</code></pre>
<p>Now it’s time to plot these results using R. Let’s first read in the table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"fstats_working/outgroupf3_europe.tsv"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can check that it worked:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Statistic   a         b     c  d NrSites Estimate_Total Estimate_Jackknife
1        F3 Han   Chuvash Mbuti NA  593124       0.053967           0.053967
2        F3 Han  Albanian Mbuti NA  593124       0.049972           0.049973
3        F3 Han  Armenian Mbuti NA  593124       0.049531           0.049531
4        F3 Han Bulgarian Mbuti NA  593124       0.050103           0.050103
5        F3 Han     Czech Mbuti NA  593124       0.050536           0.050536
6        F3 Han     Druze Mbuti NA  593124       0.048564           0.048564
  StdErr_Jackknife Z_score_Jackknife
1       0.00050668          106.5118
2       0.00049520          100.9133
3       0.00047771          103.6837
4       0.00048624          103.0419
5       0.00049261          102.5879
6       0.00046788          103.7967</code></pre>
</div>
</div>
<p>Nice, now on to plotting (here I’m using Base R for zero-dependency pain, you’re welcome!):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>order <span class="ot">&lt;-</span> <span class="fu">order</span>(d<span class="sc">$</span>Estimate_Total) <span class="co"># order the estimates for visual effect</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> d<span class="sc">$</span>Estimate_Jackknife[order]</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>xErr <span class="ot">&lt;-</span> d<span class="sc">$</span>StdErr_Jackknife[order]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">seq_along</span>(d<span class="sc">$</span>b)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">xlab =</span> <span class="st">"Z Score"</span>, <span class="at">ylab =</span> <span class="cn">NA</span>, <span class="at">yaxt =</span> <span class="st">"n"</span>, <span class="co"># plot no y-axis ticks</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim =</span> <span class="fu">c</span>(<span class="fl">0.048</span>,<span class="fl">0.06</span>),</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"F3(Han, X; Mbuti)"</span>)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the labels</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(x <span class="sc">+</span> <span class="fl">0.001</span>, y, <span class="at">labels =</span> d<span class="sc">$</span>b, <span class="at">adj=</span><span class="dv">0</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the error bars</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(x <span class="sc">-</span> xErr, y, x <span class="sc">+</span> xErr, y, <span class="at">length=</span><span class="fl">0.05</span>, <span class="at">angle=</span><span class="dv">90</span>, <span class="at">code=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fstats_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<p>As expected, the ancient samples and modern Saami are the ones with the highest allele sharing with present-day East Asians (as represented by Han) compared to many other Europeans.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Green2010" class="csl-entry" role="listitem">
Green, Richard E, Johannes Krause, Adrian W Briggs, Tomislav Maricic, Udo Stenzel, Martin Kircher, Nick Patterson, et al. 2010. <span>“A Draft Sequence of the Neandertal Genome.”</span> <em>Science</em> 328 (5979): 710–22. <a href="http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=20448178&amp;retmode=ref&amp;cmd=prlinks">http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=20448178&amp;retmode=ref&amp;cmd=prlinks</a>.
</div>
<div id="ref-Lamnidis2018" class="csl-entry" role="listitem">
Lamnidis, Thiseas C, Kerttu Majander, Choongwon Jeong, Elina Salmela, Anna Wessman, Vyacheslav Moiseyev, Valery Khartanovich, et al. 2018. <span>“Ancient Fennoscandian Genomes Reveal Origin and Spread of Siberian Ancestry in Europe.”</span> <em>Nature Communications</em> 9 (1): 5018. <a href="https://doi.org/10.1038/s41467-018-07483-5">https://doi.org/10.1038/s41467-018-07483-5</a>.
</div>
<div id="ref-Martin2015" class="csl-entry" role="listitem">
Martin, Simon H, John W Davey, and Chris D Jiggins. 2015. <span>“Evaluating the Use of <span>ABBA-BABA</span> Statistics to Locate Introgressed Loci.”</span> <em>Molecular Biology and Evolution</em> 32 (1): 244–57. <a href="https://doi.org/10.1093/molbev/msu269">https://doi.org/10.1093/molbev/msu269</a>.
</div>
<div id="ref-Patterson2012" class="csl-entry" role="listitem">
Patterson, Nick, Priya Moorjani, Yontao Luo, Swapan Mallick, Nadin Rohland, Yiping Zhan, Teri Genschoreck, Teresa Webster, and David Reich. 2012. <span>“Ancient Admixture in Human History.”</span> <em>Genetics</em> 192 (3): 1065–93. <a href="https://doi.org/10.1534/genetics.112.145037">https://doi.org/10.1534/genetics.112.145037</a>.
</div>
<div id="ref-Peter2016" class="csl-entry" role="listitem">
Peter, Benjamin M. 2016. <span>“Admixture, Population Structure, and <span>F-Statistics</span>.”</span> <em>Genetics</em> 202 (4): 1485–1501. <a href="https://doi.org/10.1534/genetics.115.183913">https://doi.org/10.1534/genetics.115.183913</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./poseidon.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./eager.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to nf-core/eager</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>