<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Martyna Molak &amp; Thiseas C. Lamnidis">
<meta name="dcterms.date" content="2023-12-07">

<title>Computational Methods for human population genetics and ancient DNA - 8&nbsp; Dimensionality reduction using PCA and MDS</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./pmrread.html" rel="next">
<link href="./fst.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./pca_mds.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Dimensionality reduction using PCA and MDS</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Computational Methods for human population genetics and ancient DNA</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Quarto_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to Quarto</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./poseidon.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fstats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentiCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Contamination estimation with AuthentiCT</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eager.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to nf-core/eager</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mobest.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">mobest - Spatiotemporal Ancestry Interpolation and Search</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fst.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Measuring population structure using Fst</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pca_mds.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Dimensionality reduction using PCA and MDS</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pmrread.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Pairwise mismatch rate (PMR) and READ relatedness inference</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yleaf.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Y-chromosomal haplogroup inference with Yleaf</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./admixfrog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Detection of archaic ancestry segments using admixfrog</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qpgraph.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Estimating Admixture Graphs with qpGraph</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#theory" id="toc-theory" class="nav-link active" data-scroll-target="#theory"><span class="header-section-number">8.1</span> Theory</a>
  <ul class="collapse">
  <li><a href="#rationale" id="toc-rationale" class="nav-link" data-scroll-target="#rationale"><span class="header-section-number">8.1.1</span> Rationale</a></li>
  <li><a href="#introduction-to-dimension-reduction" id="toc-introduction-to-dimension-reduction" class="nav-link" data-scroll-target="#introduction-to-dimension-reduction"><span class="header-section-number">8.1.2</span> Introduction to dimension reduction</a></li>
  <li><a href="#the-problem-with-missing-data" id="toc-the-problem-with-missing-data" class="nav-link" data-scroll-target="#the-problem-with-missing-data"><span class="header-section-number">8.1.3</span> The problem with missing data</a></li>
  <li><a href="#projection" id="toc-projection" class="nav-link" data-scroll-target="#projection"><span class="header-section-number">8.1.4</span> Projection</a></li>
  <li><a href="#shrinkage" id="toc-shrinkage" class="nav-link" data-scroll-target="#shrinkage"><span class="header-section-number">8.1.5</span> Shrinkage</a></li>
  </ul></li>
  <li><a href="#practice" id="toc-practice" class="nav-link" data-scroll-target="#practice"><span class="header-section-number">8.2</span> Practice</a>
  <ul class="collapse">
  <li><a href="#preparation" id="toc-preparation" class="nav-link" data-scroll-target="#preparation"><span class="header-section-number">8.2.1</span> Preparation</a></li>
  <li><a href="#run-pca" id="toc-run-pca" class="nav-link" data-scroll-target="#run-pca"><span class="header-section-number">8.2.2</span> Run PCA</a></li>
  <li><a href="#plot-pca" id="toc-plot-pca" class="nav-link" data-scroll-target="#plot-pca"><span class="header-section-number">8.2.3</span> Plot PCA</a></li>
  <li><a href="#plink-mds" id="toc-plink-mds" class="nav-link" data-scroll-target="#plink-mds"><span class="header-section-number">8.2.4</span> PLINK MDS</a></li>
  <li><a href="#compare-plots" id="toc-compare-plots" class="nav-link" data-scroll-target="#compare-plots"><span class="header-section-number">8.2.5</span> Compare plots</a></li>
  </ul></li>
  <li><a href="#conclusions" id="toc-conclusions" class="nav-link" data-scroll-target="#conclusions"><span class="header-section-number">8.3</span> Conclusions</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Dimensionality reduction using PCA and MDS</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Martyna Molak &amp; Thiseas C. Lamnidis </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 7, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="theory" class="level2 page-columns page-full" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="theory"><span class="header-section-number">8.1</span> Theory</h2>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Comparison of PCA and MDS
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<table class="table">
<colgroup>
<col style="width: 20%">
<col style="width: 23%">
<col style="width: 55%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>PCA</th>
<th>MDS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Input</td>
<td>original data matrix / similarity matrix</td>
<td>pairwise distance matrix</td>
</tr>
<tr class="even">
<td>Focus</td>
<td>captures maximum variance in data</td>
<td>preserves pairwise distances</td>
</tr>
<tr class="odd">
<td>Missing data?</td>
<td>Fill-in OR projection</td>
<td>Not an issue if using summary statistics, but this hides the uncertainty of the statistic</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<section id="rationale" class="level3 page-columns page-full" data-number="8.1.1">
<h3 data-number="8.1.1" class="anchored" data-anchor-id="rationale"><span class="header-section-number">8.1.1</span> Rationale</h3>
<p>The datasets used in human ancient DNA analysis are often extremely multidimensional, often including data from thousands of individuals, across hundreds of thousands (or millions!) of single nucleotide polymrphisms (SNPs) <span class="citation" data-cites="Mallick2023">(<a href="references.html#ref-Mallick2023" role="doc-biblioref">Mallick et al. 2023</a>)</span>. Even when choosing to summarise this genome-wide information to single statistics of genetic similarity (e.g.&nbsp;with Outgroup F3), a similarity matrix across individuals can become very large when comparing across hundreds of individuals. As the name implies, dimensionality reduction methods can reduce the number of dimensions in the underlying data, while also aiming to minimise the loss of information. The two such methods we will focus on in this tutorial are <strong>Principal Component Analysis (PCA)</strong> and <strong>Multi-Dimensional Scaling (MDS)</strong>. Both methods reveal structure within the dataset, and part of that structure is due to shared population history between individuals/populations. It is for that reason that both these methods are indispensible parts of an archaeogeneticist’s toolkit.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Mallick2023" class="csl-entry" role="listitem">
Mallick, Swapan, Adam Micco, Matthew Mah, Harald Ringbauer, Iosif Lazaridis, Iñigo Olalde, Nick Patterson, and David Reich. 2023. <span>“The Allen Ancient <span>DNA</span> Resource (<span>AADR</span>): A Curated Compendium of Ancient Human Genomes,”</span> April. <a href="https://doi.org/10.1101/2023.04.06.535797">https://doi.org/10.1101/2023.04.06.535797</a>.
</div></div></section>
<section id="introduction-to-dimension-reduction" class="level3" data-number="8.1.2">
<h3 data-number="8.1.2" class="anchored" data-anchor-id="introduction-to-dimension-reduction"><span class="header-section-number">8.1.2</span> Introduction to dimension reduction</h3>
<p>When using either of these methods, we are essentially representing the data on a new set of <strong>orthogonal</strong> axes, with its origin in the center of the data. In PCA we typically use the original data for this transformation (i.e.&nbsp;the genotype matrix), and attempt to find the axes that capture the most variation among the samples. A covariance matrix (i.e.&nbsp;a similarity matrix) is often calculated and used as a useful intermediate step in PCA. Instead, in MDS we start with a pairwise distance matrix (typically a matrix of 1-F3), and attempt to find a spatial representation that best captures the distances between points.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/pca.gif" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption class="figure-caption">A visual representation of the transformation PCA applies to a cloud of points, the range of which is represented by the blue oval. First, the data is rescaled around its own mean value, effectively moving the origin to the center of the data cloud (here shown s a red point). Then, the axes of maximal variation are discovered using linear algebra. Finally, the data is transformed to represent it along the identified axes of variation.</figcaption>
</figure>
</div>
<p>The results of both of these methods are (usually) a 2 dimensional plot in which the distances between individual points roughly correlates to the genetic distance between these individuals. Therefore, genetically similar individuals will be plotted close to one another, and further away from individuals that are more genetically dissimilar.</p>
</section>
<section id="the-problem-with-missing-data" class="level3 page-columns page-full" data-number="8.1.3">
<h3 data-number="8.1.3" class="anchored" data-anchor-id="the-problem-with-missing-data"><span class="header-section-number">8.1.3</span> The problem with missing data</h3>
<p>A recurring issue when analysing ancient DNA is the high degree of missing data (i.e.&nbsp;missingness). We often apply a minimum coverage filter to our datasets: A generally accepted rule-of-thumb for the 1240K dataset is a minimum of 15 000 covered (i.e.&nbsp;non-missing) SNPs. Another way to express this cutoff is to say that we will “happily” analyse data that is missing a genotype call in <strong>98.8%</strong> of all SNPs in the dataset! So how does this high rate of missingness affect MDS and PCA?</p>
<section id="mds" class="level4" data-number="8.1.3.1">
<h4 data-number="8.1.3.1" class="anchored" data-anchor-id="mds"><span class="header-section-number">8.1.3.1</span> MDS</h4>
<p>Missingness does not affect MDS as adversely as it does PCA, on account of the use of a pairwise distance matrix of 1-F3. This matrix will only have missing values in cases where there is no overlapping coverage between two individuals/populations used in an F3 statistic. Instead, the issue with MDS is that all F3 statistics are treated as equally reliable, regardless of their associated error bar.</p>
</section>
<section id="pca" class="level4 page-columns page-full" data-number="8.1.3.2">
<h4 data-number="8.1.3.2" class="anchored" data-anchor-id="pca"><span class="header-section-number">8.1.3.2</span> PCA</h4>
<p>Unlike MDS, PCA is severely affected by missing data. During the rescaling of the data around its own mean values, missing data is “filled-in” to the mean value (mean imputation). This can cause points to shift towards the origin by a distance relative to the degree of missingness.</p>
<p>Below is a plot of the results of PCA on a dataset of differnt worldwide populations <span class="citation" data-cites="Reich2012">(<a href="references.html#ref-Reich2012" role="doc-biblioref">Reich et al. 2012</a>)</span>. In an attempt to limit the effects of colonial admixture on the studied Native American populations, the authors masked parts of the genomes of Native Americans that matched the European or African populations in their dataset, replacing those genotypes with missing data.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Reich2012" class="csl-entry" role="listitem">
Reich, David, Nick Patterson, Desmond Campbell, Arti Tandon, Stéphane Mazieres, Nicolas Ray, Maria V Parra, et al. 2012. <span>“Reconstructing Native American Population History.”</span> <em>Nature</em> 488 (7411): 370–74.
</div></div><div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/missingness_in_PCA.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">PCA results on the same dataset based on the raw data (left), and after masking parts of the genome in Native American populations that match European populations. Due to mean imputation, masked individuals are attracted to the Origin.</figcaption>
</figure>
</div>
<p>As you can see, individuals whose genotypes were masked are shifted towards the plot’s origin. So how can we use PCA with ancient samples that have high degrees of missingness? The answer is by using a <strong>L</strong>east <strong>Sq</strong>uares <strong>Project</strong>ion, a.k.a. <code>lsqproject</code>!</p>
</section>
</section>
<section id="projection" class="level3" data-number="8.1.4">
<h3 data-number="8.1.4" class="anchored" data-anchor-id="projection"><span class="header-section-number">8.1.4</span> Projection</h3>
<p>The idea of projection is simple, and applies similarly to both PCA and MDS. In PCA, you use a subset of the dataset to calculate your axes of variation, and then apply the resulting transformation to additional data, thus projecting them onto those axes. The important detail is that the variation between projected individuals is not taken into account when deciding which the axes of maximal variation are. Similarly, in MDS you project points to the MDS space based on their distances to the points that constructed the space, disregarding the distances of the projected points to one another.</p>
<p>Below is a PCA plot calculated on present-day West Eurasian populations together with some Mesolithic hunter-gatherer individuals (in light brown). In the right side plot, the ancient individuals have been included in the calculation of the principal components, while in the left side they are projected on the principal components of the present-day West Eurasians.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/projection_comparison.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">PCA plot of ancient Mesolithic hunter-gatherers and present-day West Eurasian populations. On the right, the hunter gatherers are included in the principal component calculation, while on the left, they are projected on principal components calculated on the present-day populations only. The hunter-gatherers are part of three different groups: Eastern European hunter-gatherers (brown right-facing triangles), Scandinavian hunter-gatherers (brown diamonds), and Western European hunter-gatherers (brown half-filled circles).</figcaption>
</figure>
</div>
<p>There are two things to note here:</p>
<ul>
<li>First, comparing the placement of Eastern European hunter-gatherers (brown right-facing triangles) between the two plots, you can see that projecting these individuals does indeed provide results that are not affected by mean imputation, and thus are not shifted towards the origin.</li>
<li>Secondly, if you compare the positions of the Western European hunter-gatherers (brown half-filled circles), you will notice that projection causes these individuals to be plotted closer to present-day populations.</li>
</ul>
<p>The degree of missingness in the Western European hunter-gatherers (WHG) is relatively low, and hence the shift in their placement between the two plots is not the result of mean imputation. Instead, when projected the WHG illustrate the effects of <strong>shrinkage</strong>.</p>
</section>
<section id="shrinkage" class="level3" data-number="8.1.5">
<h3 data-number="8.1.5" class="anchored" data-anchor-id="shrinkage"><span class="header-section-number">8.1.5</span> Shrinkage</h3>
<p>Shrinkage comes in two flavours:</p>
<ul>
<li><p>The kind of shrinkage you saw with the WHGs above, is pretty intuitive. When projecting populations on axes of variation that do not capture all the variation of the projected populations, they will appear as if they have less variation than reality. This translates to the points “shrinking” towards the origin slightly. that is to say, because the WHG individuals come from a population that harboured far more genetic variation than is present within present-day West Eurasian populations, much of their true variation is “hidden” when projecting them.</p></li>
<li><p>The second kind of shrinkage (a.k.a. projection bias) arises because “samples used to calculate the PC axes”stretch” the axes” (from the <a href="https://github.com/DReichLab/EIG/tree/master/POPGEN">smartpca documentation</a>). This problem is exacerbated in datasets where the number of markers far exceeds the number of samples used for PC calculation. This is often the case in human population genomics.</p></li>
</ul>
<p>While the first shrinkage flavour can be argued to be a feature of PCA, projection bias can be a problem when trying to compare present-day populations to projected ancient populations. A demonstration of the effects of shrinkage can be seen below:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/shrinkage.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Using 10 individuals of each of the three tested populations (Yoruba, French, Han) to calculate PCs, and then projected another 10 individuals of each population reveals the effects of shrinkage on the positions of the projected individuals. In the absence of shrinkage, all points originating from the same population would be overlapping.</figcaption>
</figure>
</div>
<p>Shrinkage can be corrected by scaling the eigenvectors of the projected and/or non-projected individuals to bring them more in line with one another. Below is the same dataset as above, but ran through smartpca with the parameter <code>shrinkmode: YES</code>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/shrinkmode.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Using 10 individuals of each of the three tested populations (Yoruba, French, Han) to calculate PCs, and then projected another 10 individuals of each population. Shrinkage correction was done using ‘shrinkmode: YES’</figcaption>
</figure>
</div>
<p>As a note of caution, <code>shrinkmode: Yes</code> increases the runtime greatly. An alternative would be to identify specific present-day populations that are of interest for the ancient-to-modern comparison, and project those as well. So, for example, if we were to compare Iron Age individuals from Germany with present-day individuals from Germany, then we could decide to take out some or all present-day Germans and project those as well. That would make them fully comparable.</p>
</section>
</section>
<section id="practice" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="practice"><span class="header-section-number">8.2</span> Practice</h2>
<section id="preparation" class="level3" data-number="8.2.1">
<h3 data-number="8.2.1" class="anchored" data-anchor-id="preparation"><span class="header-section-number">8.2.1</span> Preparation</h3>
<section id="get-trident" class="level4" data-number="8.2.1.1">
<h4 data-number="8.2.1.1" class="anchored" data-anchor-id="get-trident"><span class="header-section-number">8.2.1.1</span> Get trident</h4>
<p><a href="https://www.poseidon-adna.org/#/trident">Trident</a> is a <a href="https://www.poseidon-adna.org/#/">Poseidon framework</a> data management tool. It enables downloading Poseidon packages (genomic datasets, usually including ancient individuals, comprising genome-wide SNP data) from the Poseidon server, as well as creating and manipulating such packages.</p>
<p>trident for Linux:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download Trident v1.4.0.3 binary</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/poseidon-framework/poseidon-hs/releases/download/v1.4.0.3/trident-Linux</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># rename to trident</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> trident-Linux trident</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make it executable</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">chmod</span> +x trident</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># run it</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>trident for MacOS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download Trident v1.4.0.3 binary</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-LO</span> https://github.com/poseidon-framework/poseidon-hs/releases/download/v1.4.0.3/trident-macOS</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># rename to trident</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> trident-macOS trident</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make it executable</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">chmod</span> +x trident</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># run it</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>trident for Windows:<br>
</p>
<p>Download <a href="https://github.com/poseidon-framework/poseidon-hs/releases/latest/download/trident-Windows.exe">trident-Windows.exe</a> file</p>
</section>
<section id="prepare-practise-dataset" class="level4" data-number="8.2.1.2">
<h4 data-number="8.2.1.2" class="anchored" data-anchor-id="prepare-practise-dataset"><span class="header-section-number">8.2.1.2</span> Prepare practise dataset:</h4>
<p>Here, we are downloading packages (listed in pca_mds_working/exampleData.fetchFile.txt) file from the Poseidon server into the scratch/poseidon-repository directory. The datasets come from the following publications: <a href="@Patterson2012">Patterson et al.&nbsp;2012</a>, <a href="@Lazaridis2014">Lazaridis et al.&nbsp;2014</a>, <a href="@Raghavan2014">Raghavan et al.&nbsp;2014</a>, and <a href="@JeongNatureEcologyEvolution2019">Jeong et al.&nbsp;2019</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> scratch/poseidon-repository</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This will take a few seconds to pull the data from the server</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> fetch <span class="at">-d</span> scratch/poseidon-repository <span class="at">--fetchFile</span> <span class="st">"pca_mds_working/exampleData.fetchFile.txt"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check composition of one of the downloaded packages</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> scratch/poseidon-repository/2014_LazaridisNature-4.0.2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List all groups (populations) comprised by the downloaded packages</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> list <span class="at">--groups</span> <span class="at">-d</span> scratch/poseidon-repository/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize information about  the downloaded packages</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> summarise <span class="at">-d</span> scratch/poseidon-repository</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Choose populations for the analysis (list for this exercise in "exampleData.forgeFile.txt"")</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> pca_mds_working/exampleData.forgeFile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the number of listed populations to include</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> pca_mds_working/exampleData.forgeFile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create (forge) a new repository with chosen groups from the downloaded packages</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> forge <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-d</span> scratch/poseidon-repository <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> scratch/forged_package <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-n</span> PCA_package_1 <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--forgeFile</span> pca_mds_working/exampleData.forgeFile.txt <span class="dt">\</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFormat</span> EIGENSTRAT</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The created repository comprises genomic data for: 111 modern Eurasian populations, 6 modern Native American populations, and 1 Upper Palaeolithic Siberian individual MA-1 (“Mal’ta”).</p>
</section>
</section>
<section id="run-pca" class="level3" data-number="8.2.2">
<h3 data-number="8.2.2" class="anchored" data-anchor-id="run-pca"><span class="header-section-number">8.2.2</span> Run PCA</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare parameter file for the smartpca run</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> scratch/smartpca_runs/poplist1 scratch/smartpca_runs/poplist2/</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="op">&lt;&lt;EOF</span> <span class="op">&gt;</span> scratch/smartpca_runs/poplist1/parameters.par</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="st">genotypename:   scratch/forged_package/PCA_package_1.geno   ## Genotype data</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="st">snpname:    scratch/forged_package/PCA_package_1.snp        ## SNP information</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="st">indivname:  scratch/forged_package/PCA_package_1.ind        ## Individual information</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="st">evecoutname:    scratch/smartpca_runs/poplist1/PCA_poplist1.evec           ## Eigenvectors</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="st">evaloutname:    scratch/smartpca_runs/poplist1/PCA_poplist1.eval           ## Eigenvalues</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="st">poplistname:    pca_mds_working/PCA_poplists/PCA_poplist1.txt</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="st">lsqproject: YES     ## Project individuals not included in PC calculation onto the PCs</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="st">outliermode: 2      ## Turns off automatic outlier removal.</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="st">numoutevec:  4       ## The number of eigenvectors to print per sample. Default is 10.</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="op">EOF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So prepared parameter file will cause smartpca to estimate PCs using only the individuals from the populations listed in PCA_poplist1.txt and project all the remaining individuals onto those estimated PCs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run smartpca</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">smartpca</span> <span class="at">-p</span> scratch/smartpca_runs/poplist1/parameters.par</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the output files</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> scratch/smartpca_runs/poplist1/</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> scratch/smartpca_runs/poplist1/PCA_poplist1.evec</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Adding populations (Native Americans)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look into other provided poplists</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> pca_mds_working/PCA_poplists/<span class="pp">*</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">diff</span> <span class="at">-y</span> <span class="at">--suppress-common-lines</span> pca_mds_working/PCA_poplists/PCA_poplist1.txt pca_mds_working/PCA_poplists/PCA_poplist2.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace poplist1 with poplist2 in the smartpca parameter file</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">'s/poplist1/poplist2/g'</span> scratch/smartpca_runs/poplist1/parameters.par <span class="op">&gt;</span> scratch/smartpca_runs/poplist2/parameters.par</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> scratch/smartpca_runs/poplist2/parameters.par</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rerun smartpca using poplist2 (additional populations)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">smartpca</span> <span class="at">-p</span> scratch/smartpca_runs/poplist2/parameters.par</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect smartpca output</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> scratch/smartpca_runs/poplist2/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Skipping projection (running smartpca without a poplist)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> scratch/smartpca_runs/all_pops</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 7 scratch/smartpca_runs/poplist1/parameters.par <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/poplist1/all_pops/g'</span> <span class="op">&gt;</span> scratch/smartpca_runs/all_pops/parameters.par</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> 2 scratch/smartpca_runs/poplist1/parameters.par <span class="op">&gt;&gt;</span> scratch/smartpca_runs/all_pops/parameters.par</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"maxpops: 200"</span> <span class="op">&gt;&gt;</span> scratch/smartpca_runs/all_pops/parameters.par</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"fastmode: YES"</span> <span class="op">&gt;&gt;</span> scratch/smartpca_runs/all_pops/parameters.par</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> scratch/smartpca_runs/all_pops/parameters.par</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Runtime of about 2 minutes</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="ex">smartpca</span> <span class="at">-p</span> scratch/smartpca_runs/all_pops/parameters.par</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> scratch/smartpca_runs/all_pops/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>As a result we have three PCAs:</strong></p>
<table class="table">
<colgroup>
<col style="width: 18%">
<col style="width: 48%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>data used in PC estimation</th>
<th>data projected</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>PCA_poplist1</td>
<td>Eurasians</td>
<td>Native Americans, Mal’ta</td>
</tr>
<tr class="even">
<td>PCA_poplist2</td>
<td>Eurasians, Native Americans</td>
<td>Mal’ta</td>
</tr>
<tr class="odd">
<td>PCA_all_pops</td>
<td>Eurasians, Native Americans, Mal’ta</td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="plot-pca" class="level3" data-number="8.2.3">
<h3 data-number="8.2.3" class="anchored" data-anchor-id="plot-pca"><span class="header-section-number">8.2.3</span> Plot PCA</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">'remotes'</span>)) <span class="fu">install.packages</span>(<span class="st">'remotes'</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">'janno'</span>)) remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'poseidon-framework/janno'</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Load in poplist data</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="do">## poplist1 -- Eurasian populations</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>poplist1 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="st">"pca_mds_working/PCA_poplists/PCA_poplist1.txt"</span>, <span class="at">col_names =</span> <span class="st">"Pops"</span>, <span class="at">col_types =</span> <span class="st">'c'</span>)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="do">## poplist2 -- Eurasian populations + 6 Native American populations</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>poplist2 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="st">"pca_mds_working/PCA_poplists/PCA_poplist2.txt"</span>, <span class="at">col_names =</span> <span class="st">"Pops"</span>, <span class="at">col_types =</span> <span class="st">'c'</span>)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Load in eigenvector data</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>PCA_poplist1_ev <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_fwf</span>(<span class="st">"scratch/smartpca_runs/poplist1/PCA_poplist1.evec"</span>, <span class="at">col_positions=</span>readr<span class="sc">::</span><span class="fu">fwf_widths</span>(<span class="fu">c</span>(<span class="dv">20</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">12</span>,<span class="dv">12</span>,<span class="dv">19</span>), <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">"Ind"</span>,<span class="st">"PC1"</span>,<span class="st">"PC2"</span>,<span class="st">"PC3"</span>,<span class="st">"PC4"</span>,<span class="st">"Pop"</span>)), <span class="at">col_types =</span> <span class="st">'cnnnnc'</span>, <span class="at">comment=</span><span class="st">"#"</span>)</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>PCA_poplist2_ev <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_fwf</span>(<span class="st">"scratch/smartpca_runs/poplist2/PCA_poplist2.evec"</span>, <span class="at">col_positions=</span>readr<span class="sc">::</span><span class="fu">fwf_widths</span>(<span class="fu">c</span>(<span class="dv">20</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">12</span>,<span class="dv">12</span>,<span class="dv">19</span>), <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">"Ind"</span>,<span class="st">"PC1"</span>,<span class="st">"PC2"</span>,<span class="st">"PC3"</span>,<span class="st">"PC4"</span>,<span class="st">"Pop"</span>)), <span class="at">col_types =</span> <span class="st">'cnnnnc'</span>, <span class="at">comment=</span><span class="st">"#"</span>)</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>PCA_all_pops_ev <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_fwf</span>(<span class="st">"scratch/smartpca_runs/all_pops/PCA_all_pops.evec"</span>, <span class="at">col_positions=</span>readr<span class="sc">::</span><span class="fu">fwf_widths</span>(<span class="fu">c</span>(<span class="dv">20</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">12</span>,<span class="dv">12</span>,<span class="dv">19</span>), <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">"Ind"</span>,<span class="st">"PC1"</span>,<span class="st">"PC2"</span>,<span class="st">"PC3"</span>,<span class="st">"PC4"</span>,<span class="st">"Pop"</span>)), <span class="at">col_types =</span> <span class="st">'cnnnnc'</span>, <span class="at">comment=</span><span class="st">"#"</span>)</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="do">## Finally, we load in the metadata from the forged package annotation file (janno). Here, we keep only the individual Ids, country and their Lat/Lon position.</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>metadata<span class="ot">&lt;-</span>janno<span class="sc">::</span><span class="fu">read_janno</span>(<span class="st">"scratch/forged_package/PCA_package_1.janno"</span>, <span class="at">to_janno=</span>F)<span class="sc">%&gt;%</span> <span class="fu">select</span>(Poseidon_ID, Latitude, Longitude, Country) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Longitude=</span><span class="fu">as.double</span>(Longitude), <span class="at">Latitude=</span><span class="fu">as.double</span>(Latitude))</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="do">## Finally, we add the Lat/Lon information to our datasets</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>PCA_poplist1_ev <span class="ot">&lt;-</span> <span class="fu">left_join</span>(PCA_poplist1_ev, metadata, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"Ind"</span><span class="ot">=</span><span class="st">"Poseidon_ID"</span>)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Country=</span><span class="fu">as.factor</span>(Country))</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>PCA_poplist2_ev <span class="ot">&lt;-</span> <span class="fu">left_join</span>(PCA_poplist2_ev, metadata, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"Ind"</span><span class="ot">=</span><span class="st">"Poseidon_ID"</span>)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Country=</span><span class="fu">as.factor</span>(Country))</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>PCA_all_pops_ev <span class="ot">&lt;-</span> <span class="fu">left_join</span>(PCA_all_pops_ev, metadata, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"Ind"</span><span class="ot">=</span><span class="st">"Poseidon_ID"</span>)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Country=</span><span class="fu">as.factor</span>(Country))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## First we subset the dataset to only the populations in the poplist</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>moderns_pl1 <span class="ot">&lt;-</span> PCA_poplist1_ev <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Pop <span class="sc">%in%</span> poplist1<span class="sc">$</span>Pops)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>     <span class="fu">coord_equal</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.05</span>,<span class="fl">0.05</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.05</span>,<span class="fl">0.15</span>)) <span class="sc">+</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>     <span class="fu">theme_minimal</span>()</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>(</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>moderns_pl1, <span class="do">##The input data for plotting</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2) <span class="do">## Define the x and y axis</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>        )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## We can see how genetic similarity depends on geographical location by colouring the poins by longitude or latitude</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>Lon_plot <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_pl1, <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2, <span class="at">col=</span>Longitude)) <span class="do">## Here we also define the colour of the points based on a variable</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>Lat_plot <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_pl1, <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2, <span class="at">col=</span>Latitude))</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(Lon_plot, Lat_plot, <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## As the orientation (+/-) of PC coordinates sometimes can change between runs of PCA, we use this code to ensure the same "orientation" for all users and thus enable making comparisons.</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>corner_inds_pl1 <span class="ot">&lt;-</span> moderns_pl1 <span class="sc">%&gt;%</span> <span class="fu">select</span>(Ind, PC1, PC2) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Ind <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"HGDP00607"</span>, <span class="st">"Sir50"</span>))</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_pl1<span class="sc">$</span>PC1[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_pl1<span class="sc">$</span>PC1[<span class="dv">2</span>]) { PCA_poplist1_ev <span class="ot">&lt;-</span> PCA_poplist1_ev <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">PC1=</span><span class="sc">-</span>PC1)}</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_pl1<span class="sc">$</span>PC2[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_pl1<span class="sc">$</span>PC2[<span class="dv">2</span>]) { PCA_poplist1_ev <span class="ot">&lt;-</span> PCA_poplist1_ev <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">PC2=</span><span class="sc">-</span>PC2)}</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>moderns_pl1 <span class="ot">&lt;-</span> PCA_poplist1_ev <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Pop <span class="sc">%in%</span> poplist1<span class="sc">$</span>Pops)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Let's now colour the points by country. </span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>PCA_plot_1 <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_pl1, </span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2, <span class="at">col=</span>Country), </span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent (so that aggregations of points are visible).</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>PCA_plot_1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/PCA_plot_1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Practice plot A) PCA estimated using modern Eurasian genomic data with Mal’ta projected (“PCA_poplist1”)</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Now, let's see where Mal'ta individual got projected</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>PCA_plot_1 <span class="sc">+</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>PCA_poplist1_ev <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Ind<span class="sc">==</span><span class="st">"MA1.SG"</span>), <span class="do">## Extract MA1 from the entire dataset</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2),       <span class="do">## Set the x and y axes for this set of points</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="dv">17</span>                   <span class="do">## Change shape of point to solid triangle</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"scratch/PCA_plot_1.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>PCA with Native American populations added to the analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## First we reorient the PCA</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>corner_inds_pl2 <span class="ot">&lt;-</span> PCA_poplist2_ev <span class="sc">%&gt;%</span> <span class="fu">select</span>(Ind, PC1, PC2) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Ind <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"HGDP00607"</span>, <span class="st">"Sir50"</span>))</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_pl2<span class="sc">$</span>PC1[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_pl2<span class="sc">$</span>PC1[<span class="dv">2</span>]) { PCA_poplist2_ev <span class="ot">&lt;-</span> PCA_poplist2_ev <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">PC1=</span><span class="sc">-</span>PC1)}</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_pl2<span class="sc">$</span>PC2[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_pl2<span class="sc">$</span>PC2[<span class="dv">2</span>]) { PCA_poplist2_ev <span class="ot">&lt;-</span> PCA_poplist2_ev <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">PC2=</span><span class="sc">-</span>PC2)}</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>moderns_pl2 <span class="ot">&lt;-</span> PCA_poplist2_ev <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Pop <span class="sc">%in%</span> poplist2<span class="sc">$</span>Pops)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Then we plot the output</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>PCA_plot_2 <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>     <span class="fu">coord_equal</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.05</span>,<span class="fl">0.05</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.05</span>,<span class="fl">0.15</span>)) <span class="sc">+</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>     <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_pl2 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">!=</span><span class="st">"Brazil"</span> <span class="sc">&amp;</span> Country<span class="sc">!=</span><span class="st">"Mexico"</span>), </span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2, <span class="at">col=</span>Country), </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_pl2 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">==</span><span class="st">"Brazil"</span> <span class="sc">|</span> Country<span class="sc">==</span><span class="st">"Mexico"</span>), <span class="co">#Plotting the additional American populations separately to keep colors for other countries the same between the plots</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2), </span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>       <span class="fu">geom_point</span>(</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>PCA_poplist2_ev <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Ind<span class="sc">==</span><span class="st">"MA1.SG"</span>), <span class="do">## Extract MA1 from the entire dataset</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2),       <span class="do">## Set the x and y axes for this set of points</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="dv">17</span>                   <span class="do">## Change shape of point to solid triangle</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>     )</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>PCA_plot_2</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"scratch/PCA_plot_2.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/PCA_plot_2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Practice plot B) PCA estimated using modern Eurasian and Native American genomic data with Mal’ta projected (“PCA_poplist2”)</figcaption>
</figure>
</div>
<p>And a PCA with all populations, including Mal’ta, used or estimation (no projection)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## First we reorient the PCA</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>corner_inds_ap <span class="ot">&lt;-</span> PCA_all_pops_ev <span class="sc">%&gt;%</span> <span class="fu">select</span>(Ind, PC1, PC2) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Ind <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"HGDP00607"</span>, <span class="st">"Sir50"</span>))</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_ap<span class="sc">$</span>PC1[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_ap<span class="sc">$</span>PC1[<span class="dv">2</span>]) { PCA_all_pops_ev <span class="ot">&lt;-</span> PCA_all_pops_ev <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">PC1=</span><span class="sc">-</span>PC1)}</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_ap<span class="sc">$</span>PC2[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_ap<span class="sc">$</span>PC2[<span class="dv">2</span>]) { PCA_all_pops_ev <span class="ot">&lt;-</span> PCA_all_pops_ev <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">PC2=</span><span class="sc">-</span>PC2)}</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>moderns_ap <span class="ot">&lt;-</span> PCA_all_pops_ev <span class="co">#%&gt;% filter(Pop %in% poplist2$Pops) ## Poplist 2 contains all the present-day populations.</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Then we plot the output</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>PCA_plot_ap <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>     <span class="fu">coord_equal</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.05</span>,<span class="fl">0.05</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.05</span>,<span class="fl">0.15</span>)) <span class="sc">+</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>     <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_ap <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">!=</span><span class="st">"Brazil"</span> <span class="sc">&amp;</span> Country<span class="sc">!=</span><span class="st">"Mexico"</span>), </span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2, <span class="at">col=</span>Country), </span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>moderns_ap <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">==</span><span class="st">"Brazil"</span> <span class="sc">|</span> Country<span class="sc">==</span><span class="st">"Mexico"</span>), <span class="co">#Plotting the additional American populations separately to keep colors for other countries the same between the plots</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2), </span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>PCA_all_pops_ev <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Ind<span class="sc">==</span><span class="st">"MA1.SG"</span>), <span class="do">## Extract MA1 from the entire dataset</span></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2),       <span class="do">## Set the x and y axes for this set of points</span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="dv">17</span>                   <span class="do">## Change shape of point to solid triangle</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>     )</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>PCA_plot_ap</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"scratch/PCA_plot_ap.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/PCA_plot_ap.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Practice plot C) PCA estimated using modern Eurasian and Native American, and Mal’ta genomic data with no projection (“PCA_all_pops”</figcaption>
</figure>
</div>
</section>
<section id="plink-mds" class="level3" data-number="8.2.4">
<h3 data-number="8.2.4" class="anchored" data-anchor-id="plink-mds"><span class="header-section-number">8.2.4</span> PLINK MDS</h3>
<p>Now let’s get an MDS plot for pairwise distances for the same dataset</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">##Convert package to PLINK format</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="ex">./trident</span> genoconvert <span class="at">-d</span> scratch/forged_package <span class="at">--outFormat</span> PLINK</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> scratch/forged_package/</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute pairwise distances of all individuals</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> scratch/forged_package/PCA_package_1 <span class="at">--distance-matrix</span> <span class="at">--out</span> scratch/pairwise_distances</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Read in individual IDs from MDS results</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>inds <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="st">"scratch/pairwise_distances.mdist.id"</span>, <span class="at">col_types=</span><span class="st">"cc"</span>, <span class="at">col_names=</span><span class="fu">c</span>(<span class="st">"Population"</span>, <span class="st">"Poseidon_ID"</span>))</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>inds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> janno<span class="sc">::</span><span class="fu">read_janno</span>(<span class="st">"scratch/forged_package/PCA_package_1.janno"</span>, <span class="at">to_janno=</span>F)<span class="sc">%&gt;%</span> <span class="fu">select</span>(Poseidon_ID, Latitude, Longitude, Country) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Longitude=</span><span class="fu">as.double</span>(Longitude), <span class="at">Latitude=</span><span class="fu">as.double</span>(Latitude))</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Finally, we add the Lat/Lon information to our datasets</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>inds <span class="ot">&lt;-</span> <span class="fu">left_join</span>(inds, metadata, <span class="at">by=</span><span class="st">"Poseidon_ID"</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>dist_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">scan</span>(<span class="st">"scratch/pairwise_distances.mdist"</span>), <span class="at">ncol=</span><span class="fu">nrow</span>(inds))</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dist_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>?heatmap</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first try and filter for a few populations:</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(inds<span class="sc">$</span>Population)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>indices <span class="ot">&lt;-</span> inds<span class="sc">$</span>Population <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'French'</span>, <span class="st">'Greek'</span>, <span class="st">'Nganasan'</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(indices, <span class="dv">40</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Generate a heatmap of pairwise distances</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(dist_mat[indices,indices], <span class="at">labRow =</span> inds<span class="sc">$</span>Population[indices], <span class="at">labCol =</span> inds<span class="sc">$</span>Population[indices])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr) <span class="co"># This is for the pipe operator %&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>mds_coords <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(dist_mat)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mds_coords) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"C1"</span>, <span class="st">"C2"</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>mds_coords <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">as_tibble</span>(mds_coords) <span class="sc">%&gt;%</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">bind_cols</span>(inds)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>mds_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>corner_inds <span class="ot">&lt;-</span> mds_coords <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(Poseidon_ID, C1, C2) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(Poseidon_ID <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"HGDP00607"</span>, <span class="st">"Sir50"</span>))</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds<span class="sc">$</span>C1[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds<span class="sc">$</span>C1[<span class="dv">2</span>]) { mds_coords <span class="ot">&lt;-</span> mds_coords <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">C1=</span><span class="sc">-</span>C1)}</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds<span class="sc">$</span>C2[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds<span class="sc">$</span>C2[<span class="dv">2</span>]) { mds_coords <span class="ot">&lt;-</span> mds_coords <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">C2=</span><span class="sc">-</span>C2)}</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_coords) <span class="sc">+</span> </span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>       <span class="fu">geom_point</span>(<span class="at">data=</span>mds_coords <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">!=</span><span class="st">"Brazil"</span> <span class="sc">&amp;</span> Country<span class="sc">!=</span><span class="st">"Mexico"</span>),                 <span class="fu">aes</span>(<span class="at">x=</span>C1, <span class="at">y=</span>C2, <span class="at">col=</span>Country), </span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>mds_coords <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">==</span><span class="st">"Brazil"</span> <span class="sc">|</span> Country<span class="sc">==</span><span class="st">"Mexico"</span>), <span class="co">#Plotting the additional American populations separately to keep colors for other countries the same between the plots</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>C1, <span class="at">y=</span>C2), </span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>mds_coords<span class="sc">%&gt;%</span> <span class="fu">filter</span>(Poseidon_ID<span class="sc">==</span><span class="st">"MA1.SG"</span>), <span class="do">## Extract MA1 from the entire dataset</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>C1, <span class="at">y=</span>C2),       <span class="do">## Set the x and y axes for this set of points</span></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="dv">17</span>                   <span class="do">## Change shape of point to solid triangle</span></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>     )</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"scratch/MDS_plot_ap.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/MDS_plot_ap.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Practice plot D) MDS estimated using modern Eurasian and Mal’ta genomic data (“MDS_plot_ap”)</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="do">## How does MDS compare to PCA if we restrict to the populations in poplist1?</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Read in the poplist</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>poplist1 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="st">"pca_mds_working/PCA_poplists/PCA_poplist1.txt"</span>, <span class="at">col_names =</span> <span class="st">"Pops"</span>, <span class="at">col_types =</span> <span class="st">'c'</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Filter distance matrix</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>indices_pl1 <span class="ot">&lt;-</span> inds<span class="sc">$</span>Population <span class="sc">%in%</span> poplist1<span class="sc">$</span>Pops</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>dist_mat[indices_pl1, indices_pl1]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Do MDS</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>mds_coords_pl1 <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(dist_mat[indices_pl1,indices_pl1])</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mds_coords_pl1) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"C1"</span>, <span class="st">"C2"</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>mds_coords_pl1 <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">as_tibble</span>(mds_coords_pl1) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">bind_cols</span>(inds <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(inds<span class="sc">$</span>Population <span class="sc">%in%</span> poplist1<span class="sc">$</span>Pops))</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>mds_coords_pl1</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Reorient</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>corner_inds_mds1 <span class="ot">&lt;-</span> mds_coords_pl1 <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(Poseidon_ID, C1, C2) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(Poseidon_ID <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"HGDP00607"</span>, <span class="st">"Sir50"</span>))</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_mds1<span class="sc">$</span>C1[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_mds1<span class="sc">$</span>C1[<span class="dv">2</span>]) { mds_coords_pl1 <span class="ot">&lt;-</span> mds_coords_pl1 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">C1=</span><span class="sc">-</span>C1)}</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (corner_inds_mds1<span class="sc">$</span>C2[<span class="dv">1</span>] <span class="sc">&gt;</span> corner_inds_mds1<span class="sc">$</span>C2[<span class="dv">2</span>]) { mds_coords_pl1 <span class="ot">&lt;-</span> mds_coords_pl1 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">C2=</span><span class="sc">-</span>C2)}</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_coords_pl1) <span class="sc">+</span> </span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>mds_coords_pl1 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">!=</span><span class="st">"Brazil"</span> <span class="sc">&amp;</span> Country<span class="sc">!=</span><span class="st">"Mexico"</span>), </span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>C1, <span class="at">y=</span>C2, <span class="at">col=</span>Country), </span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(<span class="at">data=</span>mds_coords_pl1 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Country<span class="sc">==</span><span class="st">"Brazil"</span> <span class="sc">|</span> Country<span class="sc">==</span><span class="st">"Mexico"</span>), <span class="co">#Plotting the additional American populations separately to keep colors for other countries the same between the plots</span></span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x=</span>C1, <span class="at">y=</span>C2), </span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>                <span class="at">alpha=</span><span class="fl">0.5</span>    <span class="do">## Makes points semi transparent.</span></span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_equal</span>()<span class="sc">+</span></span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_point</span>(</span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>mds_coords<span class="sc">%&gt;%</span> <span class="fu">filter</span>(Poseidon_ID<span class="sc">==</span><span class="st">"MA1.SG"</span>), <span class="do">## Extract MA1 from the entire dataset</span></span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>C1, <span class="at">y=</span>C2),       <span class="do">## Set the x and y axes for this set of points</span></span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="dv">17</span>                   <span class="do">## Change shape of point to solid triangle</span></span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a>     )</span>
<span id="cb33-31"><a href="#cb33-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-32"><a href="#cb33-32" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"scratch/MDS_poplist1.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/MDS_poplist1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Practice plot E) MDS estimated using modern Eurasian and Native American, and Mal’ta genomic data (“MDS_poplist1”)</figcaption>
</figure>
</div>
</section>
<section id="compare-plots" class="level3" data-number="8.2.5">
<h3 data-number="8.2.5" class="anchored" data-anchor-id="compare-plots"><span class="header-section-number">8.2.5</span> Compare plots</h3>
<p>Let’s now sum up, by directly comparing all the plots we have generated:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/pca_mds/pcamds.comparePlots.jpg" class="img-fluid figure-img"></p>
</figure>
</div>
<ol type="A">
<li>PCA estimated using modern Eurasian genomic data with Mal’ta projected (“PCA_poplist1”)</li>
<li>PCA estimated using modern Eurasian and Native American genomic data with Mal’ta projected (“PCA_poplist2”)</li>
<li>PCA estimated using modern Eurasian and Native American, and Mal’ta genomic data with no projection (“PCA_all_pops”)</li>
<li>MDS estimated using modern Eurasian, and Mal’ta genomic data (“MDS_plot_ap”)</li>
<li>MDS estimated using modern Eurasian and Native American, and Mal’ta genomic data (“MDS_poplist1”)</li>
</ol>
<p>The dataset excluding Native American data comprises much less variation that makes up PC2 and hence the PC2 variation of Eurasians in plot A is stretched up compared to these in plots B and C (including Native Americans). In plot B and C it is therefore more difficult to observe differences within Eurasians along PC2 than in plot A. The variation making PC1 is comparable between the three plots. Individual Mal’ta is more closely related to Native Americans than an average Eurasian, so without Native Americans in the dataset (A) it ends up within the Eurasian variation as there is no Native American genetic signal present that would “pull” him away from the Eurasian variation towards the American variation. When Mal’ta is included in the estimation in plot C (not only projected, like in A and B), we can observe the effects of missing data, inherent for ancient genomes, causing this individual to be pulled towards the plot’s origin.</p>
<p>In MDS the missingness in Mal’ta’s data does not affect its position as it does in PCA. Inclusion of the diverged Native American data in the estimation does cause the decrease of the distances with the Eurasian population along the C2, but the difference between plot E and plot D is not as pronounced as between B and A.</p>
</section>
</section>
<section id="conclusions" class="level2" data-number="8.3">
<h2 data-number="8.3" class="anchored" data-anchor-id="conclusions"><span class="header-section-number">8.3</span> Conclusions</h2>
<p>In PCA it is important to estimate the Eigenvalues using high-coverage samples, hence usually modern datasets, such as the 1000 Genomes Project (1kGP) or Human Genome Diversty Project (HGDP), are used for the estimation and the ancient samples are then projected onto the estimated PCs. Also, as mentioned above in the “Shrinkage” section, it is a good approach to project also some modern data if they are to be directly compared to the ancient samples.</p>
<p>While MDS will be less sensitive to the effects of missing data, it disregards the uncertainty of the underlying pairwise distance estimates.</p>
<p>It is thus the best practice to perform both analyses and compare them taking the shortcomings of each into account when interpreting the relative positions of the studied individuals/populations obtained using these methods.</p>



</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./fst.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Measuring population structure using Fst</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./pmrread.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Pairwise mismatch rate (PMR) and READ relatedness inference</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>