<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ayshin Ghalichi &amp; Clemens Schmid">
<meta name="dcterms.date" content="2023-10-19">

<title>Computational Methods for human population genetics and ancient DNA - 2&nbsp; Genotype and context data management with Poseidon</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./fstats.html" rel="next">
<link href="./Quarto_intro.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./poseidon.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Computational Methods for human population genetics and ancient DNA</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Quarto_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to Quarto</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./poseidon.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fstats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentiCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Contamination estimation with AuthentiCT</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eager.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to nf-core/eager</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mobest.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">mobest - Spatiotemporal Ancestry Interpolation and Search</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fst.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Measuring population structure using Fst</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pca_mds.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Dimensionality reduction using PCA and MDS</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pmrread.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Pairwise mismatch rate (PMR) and READ relatedness inference</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yleaf.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Y-chromosomal haplogroup inference with Yleaf</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./admixfrog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Detection of archaic ancestry segments using admixfrog</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qpgraph.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Estimating Admixture Graphs with qpGraph</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#the-components-of-poseidon" id="toc-the-components-of-poseidon" class="nav-link active" data-scroll-target="#the-components-of-poseidon"><span class="header-section-number">2.1</span> The components of Poseidon</a>
  <ul class="collapse">
  <li><a href="#the-poseidon-package" id="toc-the-poseidon-package" class="nav-link" data-scroll-target="#the-poseidon-package"><span class="header-section-number">2.1.1</span> The Poseidon package</a></li>
  <li><a href="#the-software-tools" id="toc-the-software-tools" class="nav-link" data-scroll-target="#the-software-tools"><span class="header-section-number">2.1.2</span> The software tools</a></li>
  <li><a href="#the-public-archives" id="toc-the-public-archives" class="nav-link" data-scroll-target="#the-public-archives"><span class="header-section-number">2.1.3</span> The public archives</a></li>
  <li><a href="#the-minotaur-workflow" id="toc-the-minotaur-workflow" class="nav-link" data-scroll-target="#the-minotaur-workflow"><span class="header-section-number">2.1.4</span> The Minotaur workflow</a></li>
  </ul></li>
  <li><a href="#forging-a-dataset-with-trident" id="toc-forging-a-dataset-with-trident" class="nav-link" data-scroll-target="#forging-a-dataset-with-trident"><span class="header-section-number">2.2</span> Forging a dataset with <code>trident</code></a></li>
  <li><a href="#contributing-to-the-community-archive" id="toc-contributing-to-the-community-archive" class="nav-link" data-scroll-target="#contributing-to-the-community-archive"><span class="header-section-number">2.3</span> Contributing to the community archive</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Ayshin Ghalichi &amp; Clemens Schmid </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 19, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>Poseidon is an open computational framework to enable standardised and FAIR (<span class="citation" data-cites="Wilkinson2016">Wilkinson et al. (<a href="references.html#ref-Wilkinson2016" role="doc-biblioref">2016</a>)</span>) handling of genotypes with their highly relevant context information.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Wilkinson2016" class="csl-entry" role="listitem">
Wilkinson, Mark D., Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. <span>“The <span>FAIR</span> Guiding Principles for Scientific Data Management and Stewardship.”</span> <em>Scientific Data</em> 3 (1). <a href="https://doi.org/10.1038/sdata.2016.18">https://doi.org/10.1038/sdata.2016.18</a>.
</div></div><p>It includes a well-specified <strong>data format</strong>, advanced <strong>software tools</strong>, and <strong>public, community-maintained archives</strong> to support the entire archaeogenetic research cycle, from data acquisition to management, analysis and publication.</p>
<p>Poseidon and all its components is well documented at <a href="https://www.poseidon-adna.org" class="uri">https://www.poseidon-adna.org</a>. In this tutorial we will give a brief overview and highlight two workflows in the context of Poseidon.</p>
<section id="the-components-of-poseidon" class="level2 page-columns page-full" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="the-components-of-poseidon"><span class="header-section-number">2.1</span> The components of Poseidon</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/poseidon/popgen_toolbox_schema.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption">Overview of the Poseidon framework</figcaption>
</figure>
</div>
<p>Poseidon is an entire ecosystem built on top of the data format specification of the Poseidon package.</p>
<section id="the-poseidon-package" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="the-poseidon-package"><span class="header-section-number">2.1.1</span> The Poseidon package</h3>
<p>A <a href="https://www.poseidon-adna.org/#/standard"><strong>Poseidon package</strong></a> bundles <a href="https://www.poseidon-adna.org/#/genotype_data">genotype data in EIGENSTRAT/PLINK format</a> with human- and machine-readable meta-data.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/poseidon/popgen_toolbox_package_schema.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption">The files in a Poseidon package</figcaption>
</figure>
</div>
<p>It includes sample-wise context information like spatio-temporal origin and genetic data quality in <a href="https://www.poseidon-adna.org/#/janno_details">the <code>.janno</code>-file</a>, literature in the <code>.bib</code>-file, and pointers to sequencing data in <a href="https://www.poseidon-adna.org/#/ssf_details">the <code>.ssf</code>-file</a>.</p>
<p><code>.janno</code> and <code>.ssf</code> have many predefined and specified columns, but can store arbitrary additional variables.</p>
</section>
<section id="the-software-tools" class="level3 page-columns page-full" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="the-software-tools"><span class="header-section-number">2.1.2</span> The software tools</h3>
<p><a href="https://www.poseidon-adna.org/#/trident"><strong>trident</strong></a> is a command line application to create, download, inspect and merge Poseidon packages – and therefore the central tool of the Poseidon framework. The <code>init</code> subcommand creates new packages from genotype data, <code>fetch</code> downloads them from the public archives through the Web-API, and <code>forge</code> merges and subsets them as specified. <code>list</code> gives an overview over entities in a set of packages and <code>validate</code> confirms their structural integrity.</p>
<p><a href="https://www.poseidon-adna.org/#/xerxes"><strong>xerxes</strong></a> is derived from trident and allows to directly perform various basic and experimental genomic data analyses on Poseidon packages. It implements allele sharing statistics (<span class="math inline">\(F_2\)</span>, <span class="math inline">\(F_3\)</span>, <span class="math inline">\(F_4\)</span>, <span class="math inline">\(F_{ST}\)</span>) with a flexible permutation interface.</p>
<p><a href="https://www.poseidon-adna.org/#/janno_r_package"><strong>janno</strong></a> is an R package to simplify reading .janno files into R and the popular tidyverse ecosystem (<span class="citation" data-cites="Wickham2019">Wickham et al. (<a href="references.html#ref-Wickham2019" role="doc-biblioref">2019</a>)</span>). It provides an S3 class <code>janno</code> that inherits from <code>tibble</code>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Wickham2019" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the <span class="nocase">tidyverse</span>.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div></div><p><a href="https://www.poseidon-adna.org/#/qjanno"><strong>qjanno</strong></a> is another command line tool to perform SQL queries on <code>.janno</code> files. On start-up it creates an <a href="https://www.sqlite.org/index.html">SQLite</a> database in memory and reads <code>.janno</code> files into it. It then sends any user-provided SQL query to the database server and forwards its output.</p>
</section>
<section id="the-public-archives" class="level3 page-columns page-full" data-number="2.1.3">
<h3 data-number="2.1.3" class="anchored" data-anchor-id="the-public-archives"><span class="header-section-number">2.1.3</span> The public archives</h3>
<p>The Poseidon community maintains <a href="https://www.poseidon-adna.org/#/archive_overview"><strong>public archives</strong></a> for Poseidon packages to establish a central open point of access for published archaeogenetic genotype data.</p>
<ul>
<li>The <strong>Community Archive</strong>: Author supplied per-paper packages with the genotype data published in the respective papers. Partially pre-populated from various versions of the AADR.</li>
<li>The <strong>AADR Archive</strong>: Complete and structurally unified releases of the Allen Ancient DNA Resource (<span class="citation" data-cites="Mallick2023">Mallick et al. (<a href="references.html#ref-Mallick2023" role="doc-biblioref">2023</a>)</span>) repackaged in the Poseidon package format.</li>
<li>The <strong>Minotaur Archive</strong>: Per-paper packages with genotype data reprocessed by the Minotaur workflow (see below).</li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-Mallick2023" class="csl-entry" role="listitem">
Mallick, Swapan, Adam Micco, Matthew Mah, Harald Ringbauer, Iosif Lazaridis, Iñigo Olalde, Nick Patterson, and David Reich. 2023. <span>“The Allen Ancient <span>DNA</span> Resource (<span>AADR</span>): A Curated Compendium of Ancient Human Genomes,”</span> April. <a href="https://doi.org/10.1101/2023.04.06.535797">https://doi.org/10.1101/2023.04.06.535797</a>.
</div></div><p>The data is versioned with Git and hosted on GitHub for easy co-editing and automatic structural validation.</p>
<p>It can be accessed through a <a href="https://www.poseidon-adna.org/#/web_api"><strong>Web-API</strong></a> with various endpoints at <em>server.poseidon-adna.org</em>, e.g.&nbsp;<a href="https://server.poseidon-adna.org/packages"><code>/packages</code></a> for a JSON list of packages in the community archive.</p>
<p>This API enables a little <a href="https://www.poseidon-adna.org/#/archive_explorer"><strong>Archive explorer</strong></a> web app on the Poseidon website.</p>
</section>
<section id="the-minotaur-workflow" class="level3 page-columns page-full" data-number="2.1.4">
<h3 data-number="2.1.4" class="anchored" data-anchor-id="the-minotaur-workflow"><span class="header-section-number">2.1.4</span> The Minotaur workflow</h3>
<p>The <a href="https://www.poseidon-adna.org/#/minotaur"><strong>Minotaur workflow</strong></a> is a semi-automatic workflow to reproducibly process published sequencing data from the International Nucleotide Sequence Database Collaboration (<a href="https://www.insdc.org">INSDC</a>) archives into Poseidon packages.</p>
<p>Community members can request new packages by submitting a build recipe as a Pull Request against a dedicated submission GitHub repository. This recipe is derived from a Sequencing Source File (<code>.ssf</code>), describing the sequencing data for the package and where it can be downloaded.</p>
<p>Using the recipe, the sequencing data gets processed through nf-core/eager (<span class="citation" data-cites="FellowsYates2021">Fellows Yates et al. (<a href="references.html#ref-FellowsYates2021" role="doc-biblioref">2021</a>)</span>) on computational infrastructure of MPI-EVA, using a standardised, yet flexible, set of parameters.</p>
<div class="no-row-height column-margin column-container"><div id="ref-FellowsYates2021" class="csl-entry" role="listitem">
Fellows Yates, James A., Thiseas C. Lamnidis, Maxime Borry, Aida Andrades Valtueña, Zandra Fagernäs, Stephen Clayton, Maxime U. Garcia, Judith Neukamm, and Alexander Peltzer. 2021. <span>“Reproducible, Portable, and Efficient Ancient Genome Reconstruction with Nf-Core/Eager.”</span> <em><span>PeerJ</span></em> 9 (March): e10947. <a href="https://doi.org/10.7717/peerj.10947">https://doi.org/10.7717/peerj.10947</a>.
</div></div><p>The generated genotypes, together with descriptive statistics of the sequencing data (Endogenous, Damage, Nr_SNPs, Contamination), are compiled into a Poseidon package and made available to users in the Minotaur archive.</p>
</section>
</section>
<section id="forging-a-dataset-with-trident" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="forging-a-dataset-with-trident"><span class="header-section-number">2.2</span> Forging a dataset with <code>trident</code></h2>
<p><code>forge</code> creates new Poseidon packages by extracting and merging packages, populations and individuals/samples from your Poseidon repositories. It can also work directly with your genotype data. In addition, <code>forge</code> allows merging of multiple data sets (packages), in contrast to <a href="https://github.com/DReichLab/EIG"><code>mergeit</code></a> which merges only two data sets at a time.</p>
<p><code>(-f/--forgeString)</code> can be used to query entire packages, groups or individuals. In general <code>--forgeString</code> query consists of multiple entities, inside <code>""</code> separated by <code>,</code> .</p>
<ul>
<li><p>To include all individuals in a Poseidon package, use <code>*</code> to surround the package title (e.g.&nbsp;<code>*2019_Jeong_InnerEurasia*</code>) . In cases where only genotype files are available, use the file name prefix.</p></li>
<li><p>To include certain group(s) from a Poseidon package, simply add them to the <code>-f</code> query. No specific markers are required. <code>Russia_HG_Karelia</code>. You don’t have to specify the group, as trident will search all packages for the given group.</p></li>
<li><p>To extract individuals only, surround them by <code>&lt;</code> and <code>&gt;</code>. <code>&lt;ALA026&gt;</code> . To exclude individuals add package name <code>*package*</code> and <code>&lt;individual&gt;</code> with a dash sign. <code>"*2021_Saag_EastEuropean-3.2.0*,-&lt;NIK003&gt;"</code><br>
</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trident</span> forge <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-p</span> pileupcaller.double.geno <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-d</span> 2021_Saag_EastEuropean-3.2.0 <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-d</span> 2016_FuNature-2.1.1 <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-f</span> <span class="st">"*pileupcaller.double*,Russia_AfontovaGora3,&lt;NIK003&gt;"</span> <span class="dt">\</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> testpackage <span class="dt">\</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFormat</span> EIGENSTRAT <span class="dt">\</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  /</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><a href="https://www.poseidon-adna.org/#/trident?id=the-forge-selection-language">Forge selection language</a></p>
<p><code>forge</code> has a an optional flag <code>--intersect,</code> that defines whether the genotype data from different packages should be merged with an <strong>intersect</strong> instead of the default <strong>union</strong> operation. The default is to output the union of all SNPs, by setting the additional SNPs from the other merged package as missing in the samples that did not have them originally. This option is useful for making a data set based on Human Origins (HO) SNPs for analysis like PCA and ADMIXTURE.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trident</span> forge <span class="dt">\</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--intersect</span> <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-p</span> pileupcaller.double.geno <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-d</span> 2012_PattersonGenetics-2.1.3 <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> testpackage_HO <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFormat</span> EIGENSTRAT <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  /</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In case of PCA, <code>--forgeFile</code> can be used to merge necessary populations/groups from the available packages in the community archive to create specific PCA configurations.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trident</span> forge <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-d</span> /path/to/community/archive <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--forgeFile</span> WestEurasia_poplist.txt <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> WE_PCA <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  /</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In addition, <code>--selectSnps</code> allows to provide <code>forge</code> with a SNP file in EIGENSTRAT (.snp) or PLINK (.bim) format to create a package with a specific selection. This option generates a package with exactly the SNPs listed in this file.</p>
</section>
<section id="contributing-to-the-community-archive" class="level2 page-columns page-full" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="contributing-to-the-community-archive"><span class="header-section-number">2.3</span> Contributing to the community archive</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/poseidon/recruiting.jpg" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption class="figure-caption">Poseidon needs your data as soon as it is published</figcaption>
</figure>
</div>
<p>To maintain the public data archives, specifically the community archive and the minotaur archive, Poseidon depends on work donations by an interested community.</p>
<p>Many practitioners of archaeogenetics both produce genotype data from archaological contexts and require the reference data from other publications, provided in public archives, to contextualize it.</p>
<p>If authors themselves provide high-quality, easily accessible versions of their data beyond the raw data available at the INSDC databases, they gain at least three important advantages:</p>
<ul>
<li>Their work will be easily findable and <strong>potentially cited more often</strong>.</li>
<li>They have primacy over how their <strong>data is communicated</strong>. That is, which genotypes, dates or group names they consider correct.</li>
<li>Their results for derived, genotype based analyses (PCA, F-Statistics, etc.) can be <strong>reproduced exactly</strong>.</li>
</ul>
<p>And the whole community wins, because sharing the tedious data preparation tasks empowers all researchers to achieve more in shorter time.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p><img src="img/poseidon/our_data.jpg" class="img-fluid" style="width:50.0%"></p>
</div></div><hr>
<p>This tutorial explains the main cornerstones of a workflow to add a new Poseidon package to the community archive after publishing the corresponding dataset. The process is documented in more detail in a <a href="https://www.poseidon-adna.org/#/archive_submission_guide"><strong>Submission guide</strong></a> on the website.</p>
<ol type="1">
<li><strong>Make yourself familiar</strong> with a number of core technologies. This is less daunting than it sounds, because: Superficial knowledge is sufficient and knowing them is useful beyond this particular task.</li>
</ol>
<ul>
<li>Creating and validating Poseidon packages with the <code>trident</code> tool.</li>
<li>Free and open source distributed version control with <a href="https://git-scm.com">Git</a>.</li>
<li>Collaborative working on Git projects with <a href="https://github.com">GitHub</a>.</li>
<li>Handling large files in Git using <a href="https://git-lfs.com">Git LFS</a>.</li>
</ul>
<ol start="2" type="1">
<li><strong>Create a package</strong> from your genotype data and fill it with a suitable set of meta and context information.</li>
</ol>
<ul>
<li><p><code>trident init</code> allows to wrap genotype data in a dummy Poseidon package. Imagine we had genotype data for a number of individuals in EIGENSTRAT format:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>myData.ind</strong></pre>
</div>
<pre><code>Sample1  M       ExamplePop1
Sample2  F       ExamplePop1
Sample3  M       ExamplePop2</code></pre>
</div>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>myData.snp</strong></pre>
</div>
<pre><code>           rs3094315     1        0.020130          752566 G A
          rs12124819     1        0.020242          776546 A G
          rs28765502     1        0.022137          832918 T C</code></pre>
</div>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>myData.geno</strong></pre>
</div>
<pre><code>099
922
999</code></pre>
</div>
<p>With <code>trident init -p myData.geno -o myPackage</code> we can create a basic package around this data.</p>
<pre><code>$ ls myPackage
myData.geno  myData.snp     myPackage.janno
myData.ind   myPackage.bib  POSEIDON.yml</code></pre></li>
<li><p>In a next step we modify <code>POSEIDON.yml</code>, <code>.janno</code> and <code>.bib</code> to include the context information we consider relevant. All of these files are well specified and documented, so we only demonstrate a minimal change for this example:</p>
<p>We replace the main contributor for the package.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>myPackage/POSEIDON.yml</strong></pre>
</div>
<pre><code>poseidonVersion: 2.7.1
title: myPackage
description: Empty package template. Please add a description
contributor:
- name: Clemens Schmid               #- name: Josiah Carberry
  email: clemens_schmid@eva.mpg.de   #  email: carberry@brown.edu
  orcid: 0000-0003-3448-5715         #  orcid: 0000-0002-1825-0097
packageVersion: 0.1.0
lastModified: 2023-10-18
genotypeData:
  format: EIGENSTRAT
  genoFile: myData.geno
  snpFile: myData.snp
  indFile: myData.ind
  snpSet: Other
jannoFile: myPackage.janno
bibFile: myPackage.bib</code></pre>
</div></li>
<li><p>When we applied all necessary modifications we can confirm that the package is still valid with <code>trident validate -d myPackage</code>.</p></li>
</ul>
<ol start="3" type="1">
<li><strong>Submit the package</strong> to the community archive.</li>
</ol>
<ul>
<li>To submit the package we have to create a fork of the <a href="https://github.com/poseidon-framework/community-archive">community archive repository on GitHub</a>. This requires a GitHub account.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/poseidon/fork.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Press the fork button in the top right corner to fork a repository on GitHub</figcaption>
</figure>
</div>
<ul>
<li><p>And then clone the fork to our computer, while omitting the large genotype data files. Note that this requires several setup steps to work correctly:</p>
<ul>
<li>Git has be installed for your computer (see <a href="https://git-scm.com/book/en/v2/Getting-Started-Installing-Git">here</a>)</li>
<li>You must have created an ssh key pair to connect to GitHub via ssh (see <a href="https://docs.github.com/en/authentication/connecting-to-github-with-ssh">here</a>)</li>
<li>Git LFS has to be installed (see <a href="https://git-lfs.com">here</a>) and configured for your user with <code>git lfs install</code></li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="va">GIT_LFS_SKIP_SMUDGE</span><span class="op">=</span>1 <span class="fu">git</span> clone git@github.com:<span class="op">&lt;</span>yourGitHubUserName<span class="op">&gt;</span>/community-archive.git</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>With the cloned repository on our system we can copy the files into the repositories directory and commit the changes.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>in the community-archive directory</strong></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> ../myPackage myPackage</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> add myPackage</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> commit <span class="at">-m</span> <span class="st">"added a first draft of myPackage"</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> push</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>In a last step we can open a Pull Request on GitHub from our fork to the original archive repository. Poseidon core members will take it from here.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/poseidon/pull_request.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">When you pushed to your fork, GitHub will automatically offer to “contribute” to the source repository</figcaption>
</figure>
</div>



</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./Quarto_intro.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to Quarto</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./fstats.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>