<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Suzanne Freilich">
<meta name="dcterms.date" content="2024-02-22">

<title>Computational Methods for human population genetics and ancient DNA - 10&nbsp; Y-chromosomal haplogroup inference with Yleaf</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./admixfrog.html" rel="next">
<link href="./pmrread.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./yleaf.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Y-chromosomal haplogroup inference with Yleaf</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Computational Methods for human population genetics and ancient DNA</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Quarto_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to Quarto</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./poseidon.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genotype and context data management with Poseidon</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fstats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to F3- and F4-Statistics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentiCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Contamination estimation with AuthentiCT</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eager.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to nf-core/eager</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mobest.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">mobest - Spatiotemporal Ancestry Interpolation and Search</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./fst.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Measuring population structure using Fst</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pca_mds.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Dimensionality reduction using PCA and MDS</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pmrread.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Pairwise mismatch rate (PMR) and READ relatedness inference</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yleaf.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Y-chromosomal haplogroup inference with Yleaf</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./admixfrog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Detection of archaic ancestry segments using admixfrog</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#resources" id="toc-resources" class="nav-link active" data-scroll-target="#resources"><span class="header-section-number">10.1</span> Resources</a>
  <ul class="collapse">
  <li><a href="#original-publication" id="toc-original-publication" class="nav-link" data-scroll-target="#original-publication"><span class="header-section-number">10.1.1</span> Original publication</a></li>
  <li><a href="#software-and-manual" id="toc-software-and-manual" class="nav-link" data-scroll-target="#software-and-manual"><span class="header-section-number">10.1.2</span> Software and manual</a></li>
  </ul></li>
  <li><a href="#background-to-y-chromosomal-haplogroup-assignment" id="toc-background-to-y-chromosomal-haplogroup-assignment" class="nav-link" data-scroll-target="#background-to-y-chromosomal-haplogroup-assignment"><span class="header-section-number">10.2</span> Background to Y-chromosomal haplogroup assignment</a></li>
  <li><a href="#what-is-yleaf" id="toc-what-is-yleaf" class="nav-link" data-scroll-target="#what-is-yleaf"><span class="header-section-number">10.3</span> What is Yleaf?</a>
  <ul class="collapse">
  <li><a href="#what-are-the-advantage-of-using-yleaf" id="toc-what-are-the-advantage-of-using-yleaf" class="nav-link" data-scroll-target="#what-are-the-advantage-of-using-yleaf"><span class="header-section-number">10.3.1</span> What are the advantage of using Yleaf?</a></li>
  <li><a href="#how-does-yleaf-compare-to-other-tools" id="toc-how-does-yleaf-compare-to-other-tools" class="nav-link" data-scroll-target="#how-does-yleaf-compare-to-other-tools"><span class="header-section-number">10.3.2</span> How does Yleaf compare to other tools?</a></li>
  <li><a href="#how-does-yleaf-work" id="toc-how-does-yleaf-work" class="nav-link" data-scroll-target="#how-does-yleaf-work"><span class="header-section-number">10.3.3</span> How does Yleaf work?</a></li>
  </ul></li>
  <li><a href="#installation-and-first-use" id="toc-installation-and-first-use" class="nav-link" data-scroll-target="#installation-and-first-use"><span class="header-section-number">10.4</span> Installation and first use</a></li>
  <li><a href="#usage" id="toc-usage" class="nav-link" data-scroll-target="#usage"><span class="header-section-number">10.5</span> Usage</a>
  <ul class="collapse">
  <li><a href="#options-that-are-most-relevant-for-ancient-dna-analysis" id="toc-options-that-are-most-relevant-for-ancient-dna-analysis" class="nav-link" data-scroll-target="#options-that-are-most-relevant-for-ancient-dna-analysis"><span class="header-section-number">10.5.1</span> Options that are most relevant for ancient DNA analysis</a></li>
  <li><a href="#minimum-usage-examples" id="toc-minimum-usage-examples" class="nav-link" data-scroll-target="#minimum-usage-examples"><span class="header-section-number">10.5.2</span> Minimum usage examples</a></li>
  <li><a href="#optimising-settings-for-ancient-dna-data" id="toc-optimising-settings-for-ancient-dna-data" class="nav-link" data-scroll-target="#optimising-settings-for-ancient-dna-data"><span class="header-section-number">10.5.3</span> Optimising settings for ancient DNA data</a></li>
  </ul></li>
  <li><a href="#output-and-post-processing" id="toc-output-and-post-processing" class="nav-link" data-scroll-target="#output-and-post-processing"><span class="header-section-number">10.6</span> Output and post-processing</a>
  <ul class="collapse">
  <li><a href="#the-log-file" id="toc-the-log-file" class="nav-link" data-scroll-target="#the-log-file"><span class="header-section-number">10.6.1</span> The Log file</a></li>
  <li><a href="#yleaf-haplogroup-prediction-function" id="toc-yleaf-haplogroup-prediction-function" class="nav-link" data-scroll-target="#yleaf-haplogroup-prediction-function"><span class="header-section-number">10.6.2</span> Yleaf Haplogroup prediction function</a></li>
  <li><a href="#quality-control-files" id="toc-quality-control-files" class="nav-link" data-scroll-target="#quality-control-files"><span class="header-section-number">10.6.3</span> Quality control files</a></li>
  <li><a href="#manual-inspection-and-haplotype-calling-of-ancient-dna-data-using-the-.out-file" id="toc-manual-inspection-and-haplotype-calling-of-ancient-dna-data-using-the-.out-file" class="nav-link" data-scroll-target="#manual-inspection-and-haplotype-calling-of-ancient-dna-data-using-the-.out-file"><span class="header-section-number">10.6.4</span> Manual inspection and haplotype calling of ancient DNA data using the .out file</a></li>
  <li><a href="#troubleshooting" id="toc-troubleshooting" class="nav-link" data-scroll-target="#troubleshooting"><span class="header-section-number">10.6.5</span> Troubleshooting</a></li>
  <li><a href="#private-mutations" id="toc-private-mutations" class="nav-link" data-scroll-target="#private-mutations"><span class="header-section-number">10.6.6</span> Private mutations</a></li>
  </ul></li>
  <li><a href="#follow-up-steps-for-further-verification" id="toc-follow-up-steps-for-further-verification" class="nav-link" data-scroll-target="#follow-up-steps-for-further-verification"><span class="header-section-number">10.7</span> Follow-up steps for further verification</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements"><span class="header-section-number">10.8</span> Acknowledgements</a></li>
  <li><a href="#data-availability" id="toc-data-availability" class="nav-link" data-scroll-target="#data-availability"><span class="header-section-number">10.9</span> Data availability</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Y-chromosomal haplogroup inference with Yleaf</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Suzanne Freilich </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 22, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="resources" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="resources"><span class="header-section-number">10.1</span> Resources</h2>
<section id="original-publication" class="level3" data-number="10.1.1">
<h3 data-number="10.1.1" class="anchored" data-anchor-id="original-publication"><span class="header-section-number">10.1.1</span> Original publication</h3>
<p>Ralf, Arwin, et al.&nbsp;“Yleaf: software for human Y-chromosomal haplogroup inference from next-generation sequencing data.” <em>Molecular biology and evolution</em> 35.5 (2018): 1291-1294. <span class="citation" data-cites="Ralf2018">(<a href="references.html#ref-Ralf2018" role="doc-biblioref">Ralf et al. 2018</a>)</span></p>
</section>
<section id="software-and-manual" class="level3" data-number="10.1.2">
<h3 data-number="10.1.2" class="anchored" data-anchor-id="software-and-manual"><span class="header-section-number">10.1.2</span> Software and manual</h3>
<p>Please see the <a href="https://github.com/genid/Yleaf">Yleaf-github-page</a></p>
</section>
</section>
<section id="background-to-y-chromosomal-haplogroup-assignment" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="background-to-y-chromosomal-haplogroup-assignment"><span class="header-section-number">10.2</span> Background to Y-chromosomal haplogroup assignment</h2>
<p>The human Y-chromosome haplogroup is determined by single-nucleotide polymorphisms (SNP), markers that are located in the non-recombining portion of the chromosome (NRY) which is passed down from father to son. Knowing the haplogroup can therefore be informative about biological relatedness through the paternal lineage, as well as for exploring human mobility and migration, as one can trace haplotypes’ dispersal times and geographic distributions. The mutations that accumulate over time provide us with a Y-chromosome phylogenetic tree, and a system has been developed for naming the major clades and subclades. As more SNPs are discovered, the phylogenetic tree continues to expand with the nomenclature having to continually adapt. &nbsp;</p>
</section>
<section id="what-is-yleaf" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="what-is-yleaf"><span class="header-section-number">10.3</span> What is Yleaf?</h2>
<p>Yleaf is a python-based tool for Linux for high-resolution NRY SNP calling and haplogroup inference from NGS data, independent of library and sequencing methods</p>
<section id="what-are-the-advantage-of-using-yleaf" class="level3" data-number="10.3.1">
<h3 data-number="10.3.1" class="anchored" data-anchor-id="what-are-the-advantage-of-using-yleaf"><span class="header-section-number">10.3.1</span> What are the advantage of using Yleaf?</h3>
<ul>
<li>publicly available</li>
<li>automated</li>
<li>user-friendly</li>
<li>customisable</li>
<li>versatile</li>
<li>multithreading option</li>
<li>built-in batch option</li>
<li>optimized for large NGS data sets such as whole genomes</li>
</ul>
</section>
<section id="how-does-yleaf-compare-to-other-tools" class="level3" data-number="10.3.2">
<h3 data-number="10.3.2" class="anchored" data-anchor-id="how-does-yleaf-compare-to-other-tools"><span class="header-section-number">10.3.2</span> How does Yleaf compare to other tools?</h3>
<p>Yleaf was tested on modern and ancient published datasets, and produced high concordance in NRY-SNP calling compared to well-established tools such as SAMtools/BCFtools <span class="citation" data-cites="Li2009">(<a href="references.html#ref-Li2009" role="doc-biblioref">Li et al. 2009</a>)</span>, and GATK HaplotypeCaller <span class="citation" data-cites="McKenna2010">(<a href="references.html#ref-McKenna2010" role="doc-biblioref">McKenna et al. 2010</a>)</span>. Yleaf called the lowest number of markers due to stringent options set, however it also appears to provide the highest overall accuracy</p>
</section>
<section id="how-does-yleaf-work" class="level3" data-number="10.3.3">
<h3 data-number="10.3.3" class="anchored" data-anchor-id="how-does-yleaf-work"><span class="header-section-number">10.3.3</span> How does Yleaf work?</h3>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Yleaf requires python 3.6 and currently only works on Linux</p>
</div>
</div>
<ul>
<li>Yleaf works with both raw and aligned sequencing data in a single command</li>
<li>It accepts raw fastq files, bam, cram and vcf files</li>
<li>The batch function allows you to provide a path to a folder and it will pick up all the files sharing the same extension</li>
<li>The raw reads are aligned against the reference human genome (either hg19 or hg38) with minimap2 using default settings, which creates a SAM file. (There is also a multi-threading option -t)</li>
<li>The SAM file is converted to a BAM file with SAMtools, sorted and indexed.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>For ancient DNA it is advised to rather supply Yleaf with your quality-filtered BAM files</p>
</div>
</div>
<ul>
<li>The BAM file is filtered on reads only aligning to the Y-chromosome.</li>
<li>SAMtools mpileup function creates a raw data file for variant calling.</li>
<li>Yleaf has a function embedded to perform variant calling using the generated pileup file</li>
<li>An output folder is created containing output files in the root path for each of the input files.</li>
</ul>
</section>
</section>
<section id="installation-and-first-use" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="installation-and-first-use"><span class="header-section-number">10.4</span> Installation and first use</h2>
<p>The recommended method for installing Yleaf is with conda. The github page provides easy-to-follow instructions as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># On the command line, clone the repository:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/genid/Yleaf.git</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> Yleaf</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a conda environment called yleaf:</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env create <span class="at">--file</span> environment_yleaf.yaml</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate yleaf</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Enable modification of config file in your cloned folder:</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install <span class="at">--e</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Manual installation instructions are also provided on the github page.</p>
<p>Your folder will then look like this:</p>
<p><img src="img/yleaf/Fig1.png" class="img-fluid"></p>
<p>On first use Yleaf will download reference human genome files which you will find inside the directory <em>data</em>, shown above:</p>
<p><img src="img/yleaf/Fig2.png" class="img-fluid"></p>
<p>Alternatively, you may add custom paths to your own references that you want to use in the <em>config.txt</em> file (shown in Fig.1 above), which looks like this:</p>
<p><img src="img/yleaf/Fig3.png" class="img-fluid"></p>
</section>
<section id="usage" class="level2" data-number="10.5">
<h2 data-number="10.5" class="anchored" data-anchor-id="usage"><span class="header-section-number">10.5</span> Usage</h2>
<p>To use, on your command line type:</p>
<p>Conda activate yleaf</p>
<p>Yleaf –h</p>
<p>This can be called from any directory on your system. The -h (help) flag will show you the various usage options available to adjust parameters such as quality thresholds.</p>
<p><img src="img/yleaf/Fig4.png" class="img-fluid"></p>
<section id="options-that-are-most-relevant-for-ancient-dna-analysis" class="level3" data-number="10.5.1">
<h3 data-number="10.5.1" class="anchored" data-anchor-id="options-that-are-most-relevant-for-ancient-dna-analysis"><span class="header-section-number">10.5.1</span> Options that are most relevant for ancient DNA analysis</h3>
<ul>
<li><code>-r , --reads_threshold</code>: The minimum number of reads for each base. (default=10)</li>
<li><code>-q , --quality_thresh</code>: Minimum quality for each read, integer between 10 and 40. (default=20)</li>
<li><code>-b , --base_majority</code>: The minimum percentage of a base result for acceptance, integer between 50 and 99.&nbsp; (default=90) (i.e.&nbsp;9 out of 10 reads need to agree with each other for the call to be accepted)</li>
<li><code>-dh, --draw_haplogroups</code>: Draw the predicted haplogroups in the haplogroup tree.</li>
<li><code>-pq , --prediction_quality</code>: stringent default of 0.95. Minimum quality of the prediction for it to be accepted (QC-scores)</li>
<li><code>-old , --use_old</code>: In version 3, Yleaf switched to using Yfull for the underlying tree structure of the haplogroups so predictions are slightly different to earlier versions that used ISOGG. You can use this flag to force it to use the previous method.</li>
<li><code>-aDNA</code>: recently added flag which ignores G &gt; A and C &gt; T mutations.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>It is recommended to still use Yleaf without the -aDNA option (or run it twice, with and without) so that you can manually check all mutations in case you wish to consider them in your haplogroup-calling (see further details below).</p>
</div>
</div>
</section>
<section id="minimum-usage-examples" class="level3" data-number="10.5.2">
<h3 data-number="10.5.2" class="anchored" data-anchor-id="minimum-usage-examples"><span class="header-section-number">10.5.2</span> Minimum usage examples</h3>
<p>The recommended input for ancient DNA samples is to supply bam files that you have processed with your usual quality filtering. (This could either be a BAM file that contains only reads you have extracted that align to the Y-chromosome, or you can provide your original BAM file and Yleaf will do that for you):</p>
<section id="bam-input" class="level4" data-number="10.5.2.1">
<h4 data-number="10.5.2.1" class="anchored" data-anchor-id="bam-input"><span class="header-section-number">10.5.2.1</span> BAM input</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Yleaf</span> <span class="at">-bam</span> file.bam <span class="at">-o</span> bam_output <span class="at">--rg</span> hg19</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bam-input-with-drawing-predicted-haplogroups-in-a-tree-and-showing-all-private-mutations" class="level4" data-number="10.5.2.2">
<h4 data-number="10.5.2.2" class="anchored" data-anchor-id="bam-input-with-drawing-predicted-haplogroups-in-a-tree-and-showing-all-private-mutations"><span class="header-section-number">10.5.2.2</span> BAM input with drawing predicted haplogroups in a tree and showing all private mutations</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Yleaf</span> <span class="at">-bam</span> file.bam <span class="at">-o</span> bam_output <span class="at">--rg</span> hg19 <span class="at">-dh</span> <span class="at">-p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="optimising-settings-for-ancient-dna-data" class="level3" data-number="10.5.3">
<h3 data-number="10.5.3" class="anchored" data-anchor-id="optimising-settings-for-ancient-dna-data"><span class="header-section-number">10.5.3</span> Optimising settings for ancient DNA data</h3>
<ul>
<li>Read number threshold (-r): The default for this is set to 10 – <strong>Note that</strong> <strong>this should be changed to -r 1</strong>, as some SNPs, especially in low coverage genomes, are expected to only have a single read covering them. If you run Yleaf with the default setting, you will obtain a results file but no haplogroup assignment.</li>
<li>Base quality threshold (-q): The default for this is set to 20, which would allow most bases to be accepted. You may however wish to raise this to -q 30.</li>
</ul>
<p>Examples:</p>
<section id="bam-input-1" class="level4" data-number="10.5.3.1">
<h4 data-number="10.5.3.1" class="anchored" data-anchor-id="bam-input-1"><span class="header-section-number">10.5.3.1</span> BAM input</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Yleaf</span> <span class="at">-bam</span> filename.bam <span class="at">-o</span> bam_output <span class="at">--rg</span> hg19 <span class="at">--r</span> 1 <span class="at">--q</span> 30</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="batch-option-by-providing-path-to-files" class="level4" data-number="10.5.3.2">
<h4 data-number="10.5.3.2" class="anchored" data-anchor-id="batch-option-by-providing-path-to-files"><span class="header-section-number">10.5.3.2</span> Batch option by providing path to files</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Yleaf</span> <span class="at">-bam</span> /path/to/files <span class="at">-o</span> bam_output <span class="at">--rg</span> hg19 <span class="at">--r</span> 1 <span class="at">--q</span> 30</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
</section>
<section id="output-and-post-processing" class="level2" data-number="10.6">
<h2 data-number="10.6" class="anchored" data-anchor-id="output-and-post-processing"><span class="header-section-number">10.6</span> Output and post-processing</h2>
<section id="the-log-file" class="level3" data-number="10.6.1">
<h3 data-number="10.6.1" class="anchored" data-anchor-id="the-log-file"><span class="header-section-number">10.6.1</span> The Log file</h3>
<p>After running Yleaf you will find a log file in your output directory detailing each process that was run and how long it took. As an example, here is the log file after running a single bam file:</p>
<p><img src="img/yleaf/Fig5.png" class="img-fluid"></p>
</section>
<section id="yleaf-haplogroup-prediction-function" class="level3" data-number="10.6.2">
<h3 data-number="10.6.2" class="anchored" data-anchor-id="yleaf-haplogroup-prediction-function"><span class="header-section-number">10.6.2</span> Yleaf Haplogroup prediction function</h3>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>not suitable for ancient DNA data – see below</p>
</div>
</div>
<p>You will also find the file Hg_prediction.hg, which is the final haplogroup prediction that Yleaf outputs, using the marker nomenclature from Yfull v10.01 (or if using the -old flag, ISOGG nomenclature), together with quality scores. The quality scores indicate whether the predicted haplogroup follows the expected backbone of the haplogroup tree structure and within-haplogroup tree structure. For further detailed information about these quality scores, please refer to the Yleaf documentation.</p>
<p><img src="img/yleaf/Fig6.png" class="img-fluid"></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This file is discussed here for completeness, however it is <strong>not</strong> suitable for use with ancient DNA data – for ancient DNA it is essential to carry out manual inspection of the intermediate output files provided, particularly the .out and .fmf files, as explained further below. In addition, if using present-day data, it is nevertheless recommended to perform manual inspection as prediction is not perfect. &nbsp;</p>
</div>
</div>
<p>You will also find a subdirectory which contains additional files:</p>
<p><img src="img/yleaf/Fig7.png" class="img-fluid"></p>
<p>Depending on the flags you used when running Yleaf, you may also find here additional files such as .pmu files for private mutations (see below) or .dh files for graphical output of a haplogroup tree.</p>
</section>
<section id="quality-control-files" class="level3" data-number="10.6.3">
<h3 data-number="10.6.3" class="anchored" data-anchor-id="quality-control-files"><span class="header-section-number">10.6.3</span> Quality control files</h3>
<p>The .chr file is useful for quality control and is informative about the number and percentage of mapped reads per chromosome (this comes from the output of the SAMtools command, idxstats).</p>
<p><img src="img/yleaf/Fig8.png" class="img-fluid"></p>
<p>The .info file provides general information such as total number of mapped and unmapped reads, number of markers that met the threshold or failed:</p>
<p><img src="img/yleaf/Fig9.png" class="img-fluid"></p>
<p>This file can be useful for finetuning your quality thresholds when running Yleaf. For example, this file would be helpful for identifying if you forgot to change the read number flag (-r) to 1.</p>
</section>
<section id="manual-inspection-and-haplotype-calling-of-ancient-dna-data-using-the-.out-file" class="level3" data-number="10.6.4">
<h3 data-number="10.6.4" class="anchored" data-anchor-id="manual-inspection-and-haplotype-calling-of-ancient-dna-data-using-the-.out-file"><span class="header-section-number">10.6.4</span> Manual inspection and haplotype calling of ancient DNA data using the .out file</h3>
<p>The most relevant file for performing our own manual inspection and haplotype calling of our ancient DNA data is the .out file. The default output uses the Yfull haplogroup nomenclature (terminal SNP names). Here are the first few rows:</p>
<p><img src="img/yleaf/Fig10.png" class="img-fluid"></p>
<p>The column ‘reads’ refers to the number of reads after quality filtering, while the column ‘called_per’ refers to the percentage of reads that agree with the final base call, noted in column ‘called_base’. Under ‘state’, A refers to ancestral state (as in column ‘anc’), and D to derived state (column ‘der’), while ‘depth’ indicates the depth of the hit in the YFull(v10.01) tree.</p>
<p>To assign the most derived haplogroup in ancient DNA data, the following steps should be carried out, which you can of course run or automate using your own preferred method:</p>
<ul>
<li>Sort the .out file on the “Haplogroup” column</li>
<li>Filter on derived alleles (D) in the “State” column</li>
<li>Filter out transition mutations (A-G and C-T mutations) in the “mutation” column for the most conservative approach to calling the Y-chromosome haplogroup. This is because transitions commonly represent ancient damage. However, since C-&gt;T and G-&gt;A mutations are very common and sometimes the only SNPs for many haplogroup branches, you may wish to consider them too. If your data is single-stranded, you can filter for C-&gt;T on the forward strand, and G-&gt;A on the reverse strand.</li>
</ul>
<p>After these steps, you can inspect the most derived SNP with the highest depth, which you will find towards the bottom of the table. Here are the last few rows of the sorted and filtered table. In this example, the data comes from double-stranded, non-UDG-treated libraries, and an approach was taken to filter on transversion-only SNPs:</p>
<p><img src="img/yleaf/Fig11.png" class="img-fluid"></p>
<p>In this example, we find many positions supporting a G haplogroup, with the highest depth of 23 for the haplogroup G-Y140827 according to the Yfull nomenclature (highlighted in green). (More commonly I would only find one highest-depth call however here we have three). This call is supported by additional upstream markers for the G haplogroup. If there are no other derived markers directly upstream of the most derived haplogroup, such as we see in the last four rows of this table, it suggests a spurious call due to, for example, a sequencing error, contamination, or a private mutation, and therefore you should look instead for the most derived SNP that also has upstream markers along the same branch.</p>
<p>Next, you can visit the website of Yfull (<a href="https://www.yfull.com/tree/G/" class="uri">https://www.yfull.com/tree/G/</a>) to find the path of Y-chromosome SNPs back to the root, and check if the derived SNPs make sense (this is what the automated haplogroup prediction does).</p>
<p><img src="img/yleaf/Fig12.png" class="img-fluid"></p>
<p>You can also visit the ISOGG website (https://isogg.org/tree/ ) and find the haplogroup name according to the ISOGG nomenclature. You can search the page by position or marker name from your output table. For example, the first marker name listed for the highest depth haplogroup is Z31430, which tells us it is haplogroup G2a2a1a2a2a1a:</p>
<p><img src="img/yleaf/Fig13.png" class="img-fluid"></p>
<p>If we were to search for Z6778, we would find it listed under the slightly less derived G2a2a1a2a2a. We do not find Y140827 in ISOGG, most likely because it is a more recently confirmed marker that is not included in the current version. Haplogroup names change as the tree keeps growing so it is worth looking up the most derived marker in the latest version and use the latest nomenclature. There may also be new downstream markers that are not yet included in Yleaf.</p>
<p>If you use the -old flag when running Yleaf, this outputs ISOGG haplogroup nomenclature as seen in the fourth column of the table below, which shows the last few rows (it may be slightly different from the Yfull results):</p>
<p><img src="img/yleaf/Fig14.png" class="img-fluid"></p>
<p>This way it is simpler to follow ISOGG’s “path” to see you are on the right track. Here we can see straight away that the most derived haplogroup (after dismissing the last three spurious rows) &nbsp;is G2a2a1a2a2a1 with upstream markers that lead to it. Yfull does not use such a nomenclature system because the ISOGG nomenclature could change over time with the discovery of new SNPs, and it has not been updated in a while, so using the terminal SNP names as in Yfull offers more continuity moving forward.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Where you have few markers supporting a call due to the low coverage of your data, you have the option to be conservative and report a slightly more upstream haplogroup, for example, G2, or G2a2a1a2 if you prefer.</p>
</div>
</div>
</section>
<section id="troubleshooting" class="level3" data-number="10.6.5">
<h3 data-number="10.6.5" class="anchored" data-anchor-id="troubleshooting"><span class="header-section-number">10.6.5</span> Troubleshooting</h3>
<p>An additional file, the .fmf file, is an extended .out file which includes an additional column “description”. This provides information about why a marker did not pass the criteria for haplogroup assignment, for example due to low coverage or being below the base calling threshold. This file can therefore be used to troubleshoot and optimize the settings, for example if you see that many markers were filtered out. If many markers fall below the base majority threshold, it can indicate a contaminated sample, and it could even be possible to identify the haplogroups of both contributors using the .fmf file.</p>
<p><img src="img/yleaf/Fig15.png" class="img-fluid"></p>
</section>
<section id="private-mutations" class="level3" data-number="10.6.6">
<h3 data-number="10.6.6" class="anchored" data-anchor-id="private-mutations"><span class="header-section-number">10.6.6</span> Private mutations</h3>
<p>One useful flag is the -p flag, which produces a .pmu file. This lists all identified private mutations with a minor allele frequency below 1% (this threshold can be changed with the -mr flag). It works by ignoring all SNPs that are included in Yfull and therefore associated with a haplogroup, and outputs all other SNPs found in the data. If the position is included in dbSNP (https://www.ncbi.nlm.nih.gov/snp/) it will also output its reference number (rn_no) and population frequency. This can be useful for identifying new variants in understudied populations or identifying subgroups among individuals who share the same haplogroup.</p>
<p><img src="img/yleaf/Fig16.png" class="img-fluid"></p>
</section>
</section>
<section id="follow-up-steps-for-further-verification" class="level2" data-number="10.7">
<h2 data-number="10.7" class="anchored" data-anchor-id="follow-up-steps-for-further-verification"><span class="header-section-number">10.7</span> Follow-up steps for further verification</h2>
<p>For questionable calls, it can be helpful to use a graphical interface such as <a href="https://igv.org/">IGV: Integrative Genomics Viewer</a>. Here you can load the bam and bai files (upload them via the “Tracks” option in the menu bar), ensuring you have selected the appropriate reference genome that you used for alignment (you can change it under the “Genome” option in the menu bar), and navigate to the position of the SNP along the read. Let’s say I want to take a closer look at this position from our output table:</p>
<p><img src="img/yleaf/Fig17.png" class="img-fluid"></p>
<p>I can use the search field to navigate to position chrY: 7682648 and zoom in.</p>
<p><img src="img/yleaf/Fig18.png" class="img-fluid"></p>
<p>This shows us there is one read covering this position, which has a G-&gt;C mutation, corresponding with what the output table states. We see that this SNP falls well inside the middle of the read. SNPs in the middle of a read can be interpreted as more reliable than ones at the ends of reads, which could be a result of ancient DNA damage. In such cases, especially where your data might be limited to transitions, you can check where in the read they are and if you choose, take a more conservative approach by discarding a questionable result, and instead report the haplogroup associated with the next upstream marker. Visualising the results this way can also be helpful in cases where you have two related males with slightly differing haplogroups along the same branch - in this way you can easily visualise if one of them simply lacks reads covering the more derived position that the other individual has.</p>
</section>
<section id="acknowledgements" class="level2" data-number="10.8">
<h2 data-number="10.8" class="anchored" data-anchor-id="acknowledgements"><span class="header-section-number">10.8</span> Acknowledgements</h2>
<p>The author thanks Ralf Arwin and Dion Zandstra for helpful comments and clarifications.</p>
</section>
<section id="data-availability" class="level2" data-number="10.9">
<h2 data-number="10.9" class="anchored" data-anchor-id="data-availability"><span class="header-section-number">10.9</span> Data availability</h2>
<p>The bam file used in this chapter was analysed as part of Freilich, Suzanne, et al.&nbsp;“Reconstructing genetic histories and social organisation in Neolithic and Bronze Age Croatia.” <em>Scientific Reports</em> 11.1 (2021): 16729 <span class="citation" data-cites="Freilich2021">(<a href="references.html#ref-Freilich2021" role="doc-biblioref">Freilich et al. 2021</a>)</span>, and is available from the European Nucleotide Archive <a href="http://www.ebi.ac.uk/ena">ENA</a>, under run accession <a href="https://www.ebi.ac.uk/ena/browser/view/ERR6324186">ERR6324186</a>, study&nbsp;accession number PRJEB46357.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Freilich2021" class="csl-entry" role="listitem">
Freilich, Suzanne, Harald Ringbauer, Dženi Los, Mario Novak, Dinko Tresić Pavičić, Stephan Schiffels, and Ron Pinhasi. 2021. <span>“Reconstructing Genetic Histories and Social Organisation in Neolithic and Bronze Age Croatia.”</span> <em>Scientific Reports</em> 11 (1): 1–16. <a href="https://doi.org/10.1038/s41598-021-94932-9">https://doi.org/10.1038/s41598-021-94932-9</a>.
</div>
<div id="ref-Li2009" class="csl-entry" role="listitem">
Li, Heng, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin, and 1000 Genome Project Data Processing Subgroup. 2009. <span>“The Sequence Alignment/Map Format and <span>SAMtools</span>.”</span> <em>Bioinformatics</em> 25 (16): 2078–79. <a href="https://doi.org/10.1093/bioinformatics/btp352">https://doi.org/10.1093/bioinformatics/btp352</a>.
</div>
<div id="ref-McKenna2010" class="csl-entry" role="listitem">
McKenna, Aaron, Matthew Hanna, Eric Banks, Andrey Sivachenko, Kristian Cibulskis, Andrew Kernytsky, Kiran Garimella, et al. 2010. <span>“The Genome Analysis Toolkit: A <span>MapReduce</span> Framework for Analyzing Next-Generation <span>DNA</span> Sequencing Data.”</span> <em>Genome Research</em> 20 (9): 1297–1303. <a href="https://doi.org/10.1101/gr.107524.110">https://doi.org/10.1101/gr.107524.110</a>.
</div>
<div id="ref-Ralf2018" class="csl-entry" role="listitem">
Ralf, Arwin, Diego Montiel González, Kaiyin Zhong, and Manfred Kayser. 2018. <span>“Yleaf: Software for Human <span>Y</span>-Chromosomal Haplogroup Inference from Next-Generation Sequencing Data.”</span> <em>Molecular Biology and Evolution</em> 35 (7): 1820. <a href="https://doi.org/10.1093/molbev/msy080">https://doi.org/10.1093/molbev/msy080</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./pmrread.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Pairwise mismatch rate (PMR) and READ relatedness inference</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./admixfrog.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Detection of archaic ancestry segments using admixfrog</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>